// @treeviz/gedcom-parser/dist/types/index.js
// src/types/types.ts
var RelationType = /* @__PURE__ */ ((RelationType2) => {
  RelationType2["BIOLOGICAL"] = "biological";
  RelationType2["ADOPTED"] = "adopted";
  RelationType2["FOSTER"] = "foster";
  RelationType2["SEALING"] = "sealing";
  RelationType2["BIRTH"] = "birth";
  RelationType2["STEP"] = "step";
  return RelationType2;
})(RelationType || {});
var PartnerType = /* @__PURE__ */ ((PartnerType2) => {
  PartnerType2["SPOUSE"] = "spouse";
  PartnerType2["PARTNER"] = "partner";
  PartnerType2["FRIEND"] = "friend";
  PartnerType2["SINGLE"] = "single";
  PartnerType2["OTHER"] = "other";
  PartnerType2["UNKOWN"] = "unknown";
  return PartnerType2;
})(PartnerType || {});
var Range = /* @__PURE__ */ ((Range2) => {
  Range2[Range2["Year"] = 1] = "Year";
  Range2[Range2["FiveYear"] = 5] = "FiveYear";
  Range2[Range2["Decade"] = 10] = "Decade";
  Range2[Range2["FiftyYear"] = 50] = "FiftyYear";
  Range2[Range2["Century"] = 100] = "Century";
  return Range2;
})(Range || {});

export { PartnerType, Range, RelationType };
