// @treeviz/gedcom-parser/dist/factories/index.js
import { format, isValid, parse } from 'date-fns';
import { debounce, set, get, uniqBy, unset, difference, intersectionBy, differenceBy } from 'lodash-es';

var __defProp = Object.defineProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};

// src/factories/cache-factory.ts
var defaultCacheFactory = () => {
  let cache = null;
  return {
    getItem: async () => cache,
    setItem: async (value) => {
      cache = value;
    }
  };
};
var cacheFactory = defaultCacheFactory;
var setCacheManagerFactory = (factory) => {
  cacheFactory = factory;
};
var getCacheManagerFactory = () => {
  return cacheFactory;
};
var resetCacheManagerFactory = () => {
  cacheFactory = defaultCacheFactory;
};

// src/factories/date-locale-factory.ts
var dateLocaleProvider = () => void 0;
var setDateLocaleProvider = (provider) => {
  dateLocaleProvider = provider;
};
var getDateLocale = () => {
  const result = dateLocaleProvider();
  return result;
};
var resetDateLocaleProvider = () => {
  dateLocaleProvider = () => void 0;
};

// src/factories/i18n-factory.ts
var i18nProvider = (key) => key;
var setI18nProvider = (provider) => {
  i18nProvider = provider;
};
var getI18n = () => {
  return i18nProvider;
};
var resetI18nProvider = () => {
  i18nProvider = (key) => key;
};
var i18n = {
  t: (key, options) => i18nProvider(key, options)
};

// src/kinship-translator/translators.ts
var translators_exports = {};
__export(translators_exports, {
  de: () => KinshipTranslatorDe,
  en: () => KinshipTranslatorEn,
  es: () => KinshipTranslatorEs,
  fr: () => KinshipTranslatorFr,
  hu: () => KinshipTranslatorHU
});

// src/constants/filters.ts
var FEMALE = {
  SEX: "F"
};
var MALE = {
  SEX: "M"
};
var ADOPTED = {
  PEDI: "adopted" /* ADOPTED */
};
var BIRTH = {
  PEDI: "birth" /* BIRTH */
};
var FOSTER = {
  PEDI: "foster" /* FOSTER */
};
var SEALING = {
  PEDI: "sealing" /* SEALING */
};
var STEP = {
  PEDI: "step" /* STEP */
};
var BIOLOGICAL = {
  PEDI: "biological" /* BIOLOGICAL */
};
var FRIEND = {
  PART: "friend" /* FRIEND */
};
var PARTNER = {
  PART: "partner" /* PARTNER */
};
var SPOUSE = {
  PART: "spouse" /* SPOUSE */
};

// src/utils/get-all-prop.ts
var getAllProp = (indi, prop) => {
  return prop && (indi?.get(prop)?.toList().values() ?? []).filter(Boolean) || [];
};

// src/utils/date-formatter.ts
var ACCEPTED_DATE_FORMATS = [
  "yyyy.MM.dd.",
  "yyyy-MM-dd",
  "MM/dd/yyyy"
];
var ACCEPTED_DATE_FORMATS_REGEX = new RegExp(
  `(${ACCEPTED_DATE_FORMATS.map(
    (acceptedFormat) => acceptedFormat.replace(/[^.\-/]/g, "\\d")
  ).join("|")})`,
  "gi"
);
var commonDateFormatter = (date, format2 = "yyyy", prefix = "") => {
  const formattedDate = date?.toValue(format2);
  if (!formattedDate) {
    return void 0;
  }
  return `${prefix}${formattedDate}`.trim();
};
var noteDateFormatter = (date, format2 = "yyyy", prefix = "", shortNote = true, showNote = true) => {
  const rawDate = commonDateFormatter(date, format2, "");
  if (!rawDate) {
    return void 0;
  }
  const note = date?.toNote(shortNote) ?? "";
  return `${prefix}${note && showNote ? i18n.t(`${note} {{date}}`, { date: rawDate }) : rawDate}`;
};
var marriageDateFormatter = (fam, showDays = false, showPlaces = false, shortNote = true, showNote = true, showDisplayTextIfNoDate = true, showAllEvents = false, dateFormatPattern) => {
  const fullDateFormat = i18n.t("dateFormat");
  const marriageDates = [];
  const marriagePlaces = [];
  const marriages = (fam?.toValueList().values() ?? []).flatMap((fam2) => {
    const family = fam2;
    const marrEvents = (family?.MARR?.toList().values() ?? []).filter(
      Boolean
    );
    if (marrEvents.length === 0) {
      const displayText = family?.get("DISPLAYTEXT")?.toValue() ?? "";
      if (showDisplayTextIfNoDate && displayText) {
        const displayDate = displayText.match(
          ACCEPTED_DATE_FORMATS_REGEX
        )?.[0];
        if (displayDate) {
          marriagePlaces.push(displayText.replace(displayDate, ""));
          const marriageDate = createCommonDate(
            fam2?.getGedcom(),
            void 0,
            family
          );
          marriageDate.value = displayDate;
          marriageDates.push(marriageDate);
        } else {
          marriagePlaces.push(displayText);
          marriageDates.push(void 0);
          return ["\u221E"];
        }
      }
      return [];
    }
    const eventsToProcess = showAllEvents ? marrEvents : marrEvents.slice(0, 1);
    return eventsToProcess.map((marrEvent) => {
      const marriageEventDetail = marrEvent;
      const marriageDate = marriageEventDetail?.DATE;
      const marriagePlace = showPlaces ? marriageEventDetail?.PLAC?.value : void 0;
      marriagePlaces.push(marriagePlace);
      marriageDates.push(marriageDate);
      const formattedDate = noteDateFormatter(
        marriageDate,
        showDays ? fullDateFormat : "yyyy",
        "\u221E",
        shortNote,
        showNote
      );
      return formattedDate ? formattedDate : void 0;
    });
  });
  const marriageArray = marriages.length ? marriages.filter(Boolean) : [];
  const datesArray = marriageArray.filter(Boolean);
  const placesArray = marriagePlaces.filter(Boolean);
  const dates = datesArray.join(" ").trim();
  return {
    rawArray: marriageDates,
    inArray: datesArray,
    inOrder: dates,
    places: placesArray,
    marriage: marriageArray,
    ...showPlaces ? {
      marriages,
      marriagePlaces
    } : {}
  };
};
var dateFormatter = (indi, showMarriages = false, showDays = false, showPlaces = false, shortNote = true, showNote = true, isOnStage, showAllEvents = false, dateFormatPattern) => {
  const fullDateFormat = i18n.t("dateFormat");
  const birthEvents = getAllProp(indi, "BIRT");
  const firstBirthValue = birthEvents[0];
  const deathEvents = getAllProp(indi, "DEAT");
  const firstDeathValue = deathEvents[0];
  const birth = commonDateFormatter(
    firstBirthValue?.DATE,
    showDays ? fullDateFormat : "yyyy"
  );
  const birthPlace = showPlaces ? firstBirthValue?.PLAC?.value : void 0;
  const death = commonDateFormatter(
    firstDeathValue?.DATE,
    showDays ? fullDateFormat : "yyyy"
  );
  const deathPlace = showPlaces ? firstDeathValue?.PLAC?.value : void 0;
  const marriagePlaces = [];
  const marriages = [];
  if (showMarriages || isOnStage) {
    (indi?.get("FAMS")?.toValueList().values() ?? []).forEach((fam) => {
      const famRef = fam?.ref;
      const marriageDates = marriageDateFormatter(
        famRef,
        showDays,
        showPlaces,
        shortNote,
        showNote,
        true,
        showAllEvents);
      marriages.push(...marriageDates?.marriage ?? []);
      marriagePlaces.push(...marriageDates?.marriagePlaces ?? []);
    });
  }
  const birthStrings = [];
  const birthPlaces = [];
  const eventsToProcess = showAllEvents ? birthEvents : birthEvents.slice(0, 1);
  eventsToProcess.forEach((birthEvent) => {
    const birthDate = commonDateFormatter(
      birthEvent?.DATE,
      showDays ? fullDateFormat : "yyyy"
    );
    const birthEventNote = birthEvent?.DATE?.toNote(shortNote) ?? "";
    const birthEventPlace = showPlaces ? birthEvent?.PLAC?.value : void 0;
    if (birthDate) {
      birthStrings.push(
        `*${birthEventNote && showNote ? i18n.t(`${birthEventNote} {{date}}`, {
          date: birthDate
        }) : birthDate}`
      );
    }
    birthPlaces.push(birthEventPlace);
  });
  const deathStrings = [];
  const deathPlaces = [];
  const deathEventsToProcess = showAllEvents ? deathEvents : deathEvents.slice(0, 1);
  deathEventsToProcess.forEach((deathEvent) => {
    const deathDate = commonDateFormatter(
      deathEvent?.DATE,
      showDays ? fullDateFormat : "yyyy"
    );
    const deathEventNote = deathEvent?.DATE?.toNote(shortNote) ?? "";
    const deathEventPlace = showPlaces ? deathEvent?.PLAC?.value : void 0;
    if (deathDate) {
      deathStrings.push(
        `\u2020${deathEventNote && showNote ? i18n.t(`${deathEventNote} {{date}}`, {
          date: deathDate
        }) : deathDate}`
      );
    }
    deathPlaces.push(deathEventPlace);
  });
  const birthString = birthStrings[0] ?? "";
  const deathString = deathStrings[0] ?? "";
  const marriageArray = marriages.length ? marriages.filter(Boolean) : [];
  const datesArray = !birth && !death ? marriageArray : [...birthStrings, ...marriageArray, ...deathStrings];
  const placesArray = !birth && !death ? marriagePlaces : [...birthPlaces, ...marriagePlaces, ...deathPlaces];
  const nonNullDatesArray = datesArray.filter(Boolean);
  const dates = nonNullDatesArray.join(" ").trim();
  return {
    inArray: nonNullDatesArray,
    inOrder: dates,
    places: placesArray,
    birth: birthString,
    marriage: marriageArray,
    death: deathString,
    ...showPlaces ? {
      birthPlace,
      deathPlace,
      ...showMarriages ? { marriages, marriagePlaces } : {}
    } : {},
    ...showAllEvents ? {
      births: birthStrings,
      birthPlaces,
      deaths: deathStrings,
      deathPlaces
    } : {}
  };
};

// src/utils/range.ts
var toTuple = (r) => {
  if (Array.isArray(r)) {
    const [s2, e2] = r;
    return [
      s2 === void 0 ? void 0 : Number(s2),
      e2 === void 0 ? void 0 : Number(e2)
    ];
  }
  if (r === "-" || r === "") return [void 0, void 0];
  const [s, e] = r.split("-");
  return [s ? Number(s) : void 0, e ? Number(e) : void 0];
};
var isValidRange = (range) => {
  if (!Array.isArray(range) && range !== "" && range !== "-" && !/^\d*-\d*$/.test(range)) {
    return false;
  }
  if (Array.isArray(range) && range.some((r) => typeof r === "string" && isNaN(Number(r)))) {
    return false;
  }
  return true;
};
var inRange = (year, range, trueIfNoYear = false) => {
  if (!isValidRange(range)) {
    return false;
  }
  if (!year) {
    return trueIfNoYear;
  }
  const [start = -Infinity, end = Infinity] = toTuple(range);
  if (!start && !end) {
    return true;
  }
  if (start && !end) {
    return Number(year) >= Number(start);
  }
  if (!start && end) {
    return Number(year) <= Number(end);
  }
  return Number(start) <= Number(year) && Number(year) <= Number(end);
};

// src/constants/constants.ts
var ID_REG = /^@[_a-zA-Z0-9]+@/;
var ID_GETTER_REG = /^(?<at>@)?(?<letter>[A-Z])?/i;
var List = class _List {
  // length = 0;
  constructor(items) {
    this.items = {};
    this.isListable = true;
    this.marker = void 0;
    if (items) {
      this.concat(items);
    }
  }
  get length() {
    return this.values().length;
  }
  has(indi) {
    return !!this.item(
      typeof indi === "string" ? indi : indi?.id ?? ""
    );
  }
  get(name) {
    const propList = new _List();
    this.entries().forEach(([key, item]) => {
      propList.item(key, item.get(name));
    });
    return propList;
  }
  getIf(name, condition, name2) {
    const propList = new _List();
    this.entries().forEach(([key, item]) => {
      const passedItem = item.getIf(name, condition, name2);
      if (passedItem) {
        propList.item(key, passedItem);
      }
    });
    return propList;
  }
  keys() {
    return Object.keys(this.items);
  }
  values() {
    return Object.values(this.items);
  }
  entries() {
    return Object.entries(this.items);
  }
  setLength(_type) {
  }
  first() {
    return this.index(0);
  }
  last() {
    return this.index(this.keys().length - 1);
  }
  index(index) {
    const keys = this.keys();
    const key = keys[index];
    return key && this.items[key];
  }
  item(name, value) {
    if (arguments.length > 1) {
      set(this.items, name, value);
      this.setLength();
    }
    return get(this.items, name);
  }
  removeItem(name) {
    unset(this.items, name);
    this.setLength();
  }
  exclude(excludedList) {
    excludedList.keys().forEach((itemId) => {
      delete this.items[itemId];
    });
    this.setLength();
    return this;
  }
  concat(items) {
    if (Array.isArray(items)) {
      items.forEach((item) => {
        this.append(item);
      });
    } else {
      Object.assign(this.items, items ?? {});
    }
    this.setLength();
    return this;
  }
  merge(mergedList) {
    this.concat(mergedList?.items);
    return this;
  }
  intersection(mergedList) {
    const inter = intersectionBy(
      this.values(),
      mergedList?.values() ?? [],
      "id"
    );
    return new _List(inter);
  }
  difference(mergedList) {
    const inter = differenceBy(
      this.values(),
      mergedList?.values() ?? [],
      "id"
    );
    return new _List(inter);
  }
  prepend(item) {
    if (item.id) {
      let increase = true;
      if (this.items[item.id]) {
        delete this.items[item.id];
        increase = false;
      }
      this.items = Object.assign(
        {
          [item.id]: item
        },
        this.items
      );
      if (increase) {
        this.setLength("increase");
      }
    }
    return this;
  }
  append(item) {
    if (item.id) {
      let increase = true;
      if (this.items[item.id]) {
        delete this.items[item.id];
        increase = false;
      }
      this.items[item.id] = item;
      if (increase) {
        this.setLength("increase");
      }
    }
    return this;
  }
  delete(item) {
    if (item.id && this.items[item.id]) {
      delete this.items[item.id];
      this.setLength("decrease");
    }
    return this;
  }
  copy(ClassName = _List) {
    const newList = new ClassName();
    Object.assign(newList.items, this.items);
    newList.setLength();
    return newList;
  }
  except(item, ClassName = _List) {
    return this.copy(ClassName).delete(item);
  }
  forEach(iterate) {
    this.entries().forEach(([key, item], index) => {
      iterate(item, key, index);
    });
  }
  map(iterate) {
    return this.entries().map(([key, item], index) => {
      return iterate(item, key, index);
    });
  }
  reduce(iterate, initialValue) {
    return this.entries().reduce((acc, [key, item], index) => {
      return iterate(acc, item, key, index);
    }, initialValue);
  }
  filter(filters, ClassName = _List) {
    const newList = new ClassName();
    const isIterator = typeof filters === "function";
    this.entries().forEach(([itemId, item], index) => {
      if (isIterator) {
        filters(item, itemId, index) && newList.item(itemId, item);
        return;
      }
      if (!Object.keys(filters).length || Object.entries(filters).every(([key, value]) => {
        let itemValue;
        if (key === "id") {
          itemValue = item.toProp("id");
        } else {
          itemValue = item.toProp(key)?.toValue();
        }
        if (Array.isArray(value)) {
          return value.includes(itemValue);
        }
        return itemValue === value;
      })) {
        newList.item(itemId, item);
      }
    });
    return newList;
  }
  find(filters, _ClassName = _List) {
    let foundItem;
    const isIterator = typeof filters === "function";
    this.entries().find(([itemId, item], index) => {
      if (isIterator) {
        if (filters(item, itemId, index)) {
          foundItem = item;
          return true;
        }
        return false;
      }
      if (!Object.keys(filters).length || Object.entries(filters).every(([key, value]) => {
        let itemValue;
        if (key === "id") {
          itemValue = item.toProp("id");
        } else {
          itemValue = item.toProp(key)?.toValue();
        }
        if (Array.isArray(value)) {
          return value.includes(itemValue);
        }
        return itemValue === value;
      })) {
        foundItem = item;
        return true;
      }
      return false;
    });
    return foundItem;
  }
  orderBy(orders, ClassName = _List) {
    const newList = new ClassName();
    const isIterator = typeof orders === "function";
    if (!isIterator && (!orders || !Object.keys(orders).length)) {
      return newList.merge(this);
    }
    const sortedItems = this.entries().toSorted(
      ([itemAId, itemA], [itemBId, itemB]) => {
        if (isIterator) {
          return orders(itemA, itemAId, itemB, itemBId);
        }
        const [key, { direction, getter }] = Object.entries(
          typeof orders === "string" ? { [orders]: { direction: "ASC" } } : orders
        )[0];
        const rawA = itemA.get(key);
        const rawB = itemB.get(key);
        let valueA = rawA?.toValue();
        let valueB = rawB?.toValue();
        if (getter && typeof getter === "function") {
          valueA = getter(valueA, rawA);
          valueB = getter(valueB, rawB);
        }
        let sortValue = 0;
        if (!valueA && valueA !== 0 && !valueB && valueB !== 0) {
          sortValue = 0;
        } else if (!valueB && valueB !== 0) {
          sortValue = 1;
        } else if (!valueA && valueA !== 0) {
          sortValue = -1;
        } else if (valueA < valueB) {
          sortValue = -1;
        } else if (valueA > valueB) {
          sortValue = 1;
        }
        return sortValue * (direction === "DESC" ? -1 : 1);
      }
    );
    sortedItems.forEach(([key, value]) => {
      newList.item(key, value);
    });
    return newList;
  }
  groupBy(groups, sort, minOccurance = -1, lessOccuranceLabel = "Appearing less than 5 times", ClassName = _List) {
    const groupped = {};
    const isIterator = typeof groups === "function";
    if (!isIterator && (!groups || !Object.keys(groups).length)) {
      groupped.Unknown = new ClassName().merge(this);
      groupped.Unknown.marker = "Unknown";
      return groupped;
    }
    this.forEach((item, itemId) => {
      const group = [];
      if (isIterator) {
        const newGroups = groups(item, itemId);
        group.push(
          ...Array.isArray(newGroups) ? newGroups : [newGroups]
        );
      } else {
        const [key, { getter }] = Object.entries(
          typeof groups === "string" ? { [groups]: {} } : groups
        )[0];
        const raw = item.get(key)?.index(0);
        let value = raw?.toValue();
        if (getter && typeof getter === "function") {
          value = getter(value, raw);
        }
        if (value) {
          group.push(value);
        }
      }
      if (!group.length) {
        group.push("Unknown");
      }
      group.forEach((g) => {
        const newGroup = typeof g === "string" ? { group: g } : g;
        if (!groupped[newGroup.group]) {
          groupped[newGroup.group] = new ClassName();
          groupped[newGroup.group].marker = newGroup.marker ?? newGroup.group;
        }
        groupped[newGroup.group]?.append(item);
      });
    });
    if (sort) {
      return Object.entries(groupped).toSorted(([ak, a], [bk, b]) => {
        if (typeof sort === "function") {
          return sort(a, ak, b, bk);
        }
        if (ak === "Unknown") {
          return 1;
        }
        if (bk === "Unknown") {
          return -1;
        }
        if (sort === "group") {
          return ak.localeCompare(bk);
        }
        return (b?.length ?? 0) - (a?.length ?? 0);
      }).reduce((acc, [key, value]) => {
        if (minOccurance > -1 && (value?.length ?? 0) < minOccurance) {
          const otherKey = lessOccuranceLabel;
          acc[otherKey] = acc[otherKey] && value ? acc[otherKey].merge(value) : value;
        } else {
          acc[key] = value;
        }
        return acc;
      }, {});
    }
    return groupped;
  }
  findIndex(item) {
    if (!item?.id) {
      return -1;
    }
    return this.keys().findIndex((key) => key === item.id);
  }
  getItems() {
    return this.items;
  }
  toJson(tag, options) {
    const json = this.toObject(tag, options);
    return JSON.stringify(json);
  }
  toObject(tag, options) {
    const json = {};
    this.entries().forEach(([_, item]) => {
      const validTag = getValidTag(tag);
      if (json[validTag]) {
        if (!Array.isArray(json[validTag])) {
          json[validTag] = [json[validTag]];
        }
        json[validTag].push(
          item.toObject(validTag, options)
        );
      } else {
        json[validTag] = item.toObject(
          validTag,
          options
        );
      }
    });
    return json;
  }
  toGedcom(tag, level = 0, options) {
    const gedcom = this.toGedcomLines(tag, level, options);
    return gedcom.join("\n");
  }
  toGedcomLines(tag, level = 0, options) {
    const gedcom = [];
    this.entries().forEach(([_, item]) => {
      const validTag = getValidTag(tag);
      const validKey = item.id;
      const validValue = validKey ? validTag : item.exportValue();
      gedcom.push(
        `${level} ${validKey || validTag}${validValue ? ` ${validValue}` : ""}`
      );
      gedcom.push(
        ...item.toGedcomLines(
          validTag,
          level + 1,
          options
        )
      );
    });
    return gedcom;
  }
  toValue() {
    const propList = new _List();
    this.entries().forEach(([key, item]) => {
      propList.item(key, item?.toValue());
    });
    return propList;
  }
  // @deprecated use get instead
  toProp(tag) {
    return this.get(tag);
  }
  toList() {
    return new _List().concat(this.items);
  }
  toValueList() {
    const newList = new _List();
    this.values().forEach((value) => {
      value.value && newList.item(value.value, value);
    });
    return newList;
  }
};

// src/classes/common.ts
var Common = class _Common {
  constructor(gedcom, id, main, parent) {
    this.isListable = true;
    this._gedcom = gedcom;
    this._main = main;
    this._parent = parent;
    this.initUniqueId();
    if (id) {
      this._id = id;
    } else {
      delete this._id;
    }
  }
  set type(type) {
    if (type) {
      this.uniqueId = this.uniqueId?.replace(/^[^:]+/, `~${type}`);
    }
    this._type = type;
  }
  get type() {
    return this._type;
  }
  initUniqueId() {
    if (this._gedcom) {
      this.uniqueId = `~Common:${this._gedcom.refcount++}`;
    }
  }
  set uniqueId(id) {
    const currentId = this._uniqueId;
    if (this._gedcom?.reflist && id) {
      this._uniqueId = id;
      if (currentId && this._gedcom.reflist[currentId]) {
        delete this._gedcom.reflist[currentId];
      }
      this._gedcom.reflist[id] = this;
    }
  }
  get uniqueId() {
    return this._uniqueId;
  }
  set id(id) {
    this._id = idGetter(id);
  }
  get id() {
    return idGetter(this._id);
  }
  set value(value) {
    this._value = value;
  }
  get value() {
    return this._value;
  }
  /**
   * Set the gedcom reference for this object
   * @param gedcom - The GedComType to set as the gedcom reference
   */
  setGedcom(gedcom) {
    this._gedcom = gedcom;
  }
  // avoid to override
  get originalValue() {
    return this._value;
  }
  get ref() {
    if (typeof this.value === "string" && isId(this.value) && this.refType) {
      return this._gedcom?.[getListTag(this.refType)]?.item(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.value
      );
    }
    return this;
  }
  get main() {
    return this._main;
  }
  get parent() {
    return this._parent;
  }
  addRef(refNode) {
    if (typeof refNode.value === "string" && isId(refNode.value) && refNode.refType) {
      if (!this._refs) {
        this._refs = new List();
      }
      this._refs.item(refNode.value, refNode);
    }
  }
  getRefs() {
    return this._refs;
  }
  exportValue() {
    return this.value;
  }
  removeValue() {
    delete this._value;
  }
  set(name, value) {
    if (typeof value === "string") {
      const usedValue = createCommon(this._gedcom, void 0, this.main);
      usedValue.value = value;
      set(this, name, usedValue);
    } else {
      set(this, name, value);
    }
    return get(this, name);
  }
  assign(name, value, unique = false) {
    let curValue = this.get(name);
    if (curValue?.isListable) {
      if (curValue instanceof _Common) {
        curValue = new List().concat({ ...[curValue] });
        this.set(name, curValue);
      }
      const values = [
        ...curValue.values(),
        ...value instanceof List ? value.values() : [value]
      ];
      curValue.concat({
        ...!unique ? values : uniqBy(
          values,
          (value2) => value2 instanceof _Common ? value2?.value : void 0
        )
      });
    } else {
      this.set(name, value);
    }
    return get(this, name);
  }
  get(name) {
    if (!name.includes(".")) {
      return get(this, name);
    }
    const nameParts = name.split(".");
    let last = this;
    for (const part of nameParts) {
      const indexPart = Number(part);
      const itemIndex = part.match(/items\[(?<itemIndex>\d+)\]/)?.groups?.itemIndex;
      const itemId = part.match(/items\[(?<itemId>@[A-Z]+\d+@)\]/)?.groups?.itemId;
      const usedPart = itemId || itemIndex;
      if (last instanceof List) {
        if (!isNaN(indexPart)) {
          last = last.index(indexPart);
        } else {
          if (usedPart) {
            last = last.item(usedPart);
          } else {
            last = last?.index(0)?.get(part);
          }
        }
      } else {
        const usedIndex = Number(itemIndex ?? indexPart);
        if (!isNaN(usedIndex)) {
          last = last?.toList().index(usedIndex);
        } else {
          last = last?.get(itemId ?? part);
        }
      }
      if (!last) {
        break;
      }
    }
    return last;
  }
  remove(name) {
    unset(this, name);
  }
  getGedcom() {
    return this._gedcom;
  }
  getIf(name, condition, name2) {
    const nameParts = name.split(".");
    const name2Parts = name2.split(".");
    if (nameParts.length !== name2Parts.length) {
      throw new Error("Tag pathes must be of the same depth");
    }
    if (nameParts.length > 1 && nameParts[0] !== name2Parts[0]) {
      throw new Error("Final tags must be siblings");
    }
    const valueCommon = this.get(name);
    if (valueCommon instanceof _Common) {
      return valueCommon?.toValue() !== condition ? void 0 : this.get(name2);
    }
    const valueCommon2 = this.get(name2)?.toList();
    const ifList = new List();
    valueCommon?.forEach((item, key) => {
      const pair = valueCommon2?.item(key);
      ifList.item(key, item?.toValue() !== condition ? void 0 : pair);
    });
    return ifList;
  }
  toString() {
    return this.value?.toString() || "";
  }
  toValue() {
    return this.value;
  }
  toProp(tag) {
    const prop = this.get(tag);
    return prop;
  }
  toList() {
    return new List().concat(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.id ? { [this.id]: this } : { ...[this] }
    );
  }
  index(i) {
    return this.toList()?.index(i);
  }
  toValueList() {
    return new List().concat(
      this.value ? { [this.value]: this } : (
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        { ...[this] }
      )
    );
  }
  standardizeObject(tag, options) {
    if (tag === "OBJE" && !options?.original && options?.obje?.standardize && options?.obje?.namespace && "standardizeMedia" in this) {
      const standardize = this.standardizeMedia;
      return standardize.call(
        this,
        options.obje.namespace,
        options.obje.override
      );
    }
  }
  toJson(tag, options) {
    const json = this.toObject(tag, options);
    return JSON.stringify(json);
  }
  toObject(tag, options) {
    this.standardizeObject(tag, options);
    const validKeys = getValidKeys(this);
    let json = {};
    validKeys.forEach((key) => {
      if (key === "id" && this.id !== void 0) {
        json.id = this.id;
      } else if (key === "_id" && this._id !== void 0) {
        json.id = this._id;
      } else if (key === "value" && this.value !== void 0) {
        json.value = this.value;
      } else if (key === "_value" && this._value !== void 0) {
        json.value = this._value;
      } else {
        const validKey = key;
        const prop = this.get(validKey);
        if (typeof prop?.toObject === "function") {
          if (prop instanceof _Common) {
            json[validKey] = prop.toObject(validKey, options);
          } else {
            json = {
              ...json,
              ...prop.toObject(validKey, options)
            };
          }
        }
      }
    });
    return json;
  }
  merge(other, override = false, avoidKeys = []) {
    if (!other) {
      return this;
    }
    const validKeys = getValidKeys(other);
    validKeys.forEach((validKey) => {
      if (avoidKeys.includes(validKey)) {
        return;
      }
      const current = this.get(validKey);
      const mergeValue = other.get(validKey);
      if (mergeValue) {
        if (current instanceof _Common || current instanceof List) {
          this.assign(validKey, mergeValue, true);
        } else if (!current || override && validKey !== "id" && validKey !== "_id") {
          this.set(validKey, mergeValue);
        }
      }
    });
    return this;
  }
  clone(newId = false, avoidKeys = []) {
    const ctor = this.constructor;
    const instance = new ctor(
      this._gedcom,
      this._id,
      this._main,
      this._parent
    );
    if (newId && this._id) {
      instance.id = `IFT${this._id.replace(/[^0-9]/g, "")}`;
    }
    const cloned = createProxy(instance);
    cloned.merge(this, true, avoidKeys);
    return cloned;
  }
  // fromGedcom<T = typeof this>(value: string) {
  fromGedcom(_value) {
  }
  toGedcom(tag, level = 0, options) {
    const gedcom = this.toGedcomLines(tag, level, options);
    return gedcom.join("\n");
  }
  toGedcomLines(tag, level = 0, options) {
    this.standardizeObject(tag, options);
    const validKeys = getValidKeys(this);
    const gedcom = [];
    validKeys.forEach((key) => {
      const validKey = key;
      const prop = this.get(validKey);
      if (typeof prop?.toGedcomLines === "function") {
        if (prop instanceof _Common) {
          const value = prop.exportValue();
          gedcom.push(
            `${level} ${validKey}${value ? ` ${value}` : ""}`
          );
          gedcom.push(
            ...prop.toGedcomLines(validKey, level + 1, options)
          );
        } else {
          gedcom.push(
            ...prop.toGedcomLines(validKey, level, options)
          );
        }
      }
    });
    return gedcom;
  }
  isGenoPro() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sour = get(head, "SOUR.value");
    return !!sour?.toLowerCase()?.startsWith("genopro");
  }
  isAhnenblatt() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sour = get(head, "SOUR.value");
    return !!sour?.toLowerCase()?.startsWith("ahnenblatt");
  }
  isGeni() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sour = get(head, "SOUR.value");
    return !!sour?.toLowerCase()?.startsWith("geni");
  }
  isAncestry() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sour = get(head, "SOUR.value");
    return !!sour?.toLowerCase()?.startsWith("ancestry");
  }
  isMyHeritage() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sour = get(head, "SOUR.value");
    return !!sour?.toLowerCase()?.startsWith("myheritage");
  }
  /**
   * Get the source type as a string (for prefixing tree IDs and names)
   * Returns the detected source type or undefined if unknown
   */
  getSourceType() {
    if (this.isAncestry()) return "Ancestry";
    if (this.isMyHeritage()) return "MyHeritage";
    if (this.isFamilySearch()) return "FamilySearch";
    if (this.isGNO2GED()) return "GNO2GED";
    if (this.isGenoPro()) return "GenoPro";
    if (this.isAhnenblatt()) return "Ahnenblatt";
    if (this.isGeni()) return "Geni";
    return void 0;
  }
  isFamilySearch() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sourName = get(head, "SOUR.NAME.value");
    return sourName === "FamilySearch API";
  }
  isGNO2GED() {
    const head = get(this, "HEAD") || get(this.getGedcom(), "HEAD");
    const sour = get(head, "SOUR.value");
    return sour === "GNO2GED";
  }
  getAncestryTreeId() {
    const path = "HEAD.SOUR._TREE.RIN.value";
    return get(this, path) || get(this.getGedcom(), path);
  }
  getMyHeritageTreeId() {
    const path = "HEAD._EXPORTED_FROM_SITE_ID.value";
    return get(this, path) || get(this.getGedcom(), path);
  }
  getTreeId() {
    if (this?.isAncestry()) {
      const id = this.getAncestryTreeId();
      if (id !== void 0) return id;
    }
    if (this?.isMyHeritage()) {
      const id = this.getMyHeritageTreeId();
      if (id !== void 0) return id;
    }
    if (this?.isFamilySearch()) {
      const id = this.getFamilySearchTreeId();
      if (id !== void 0) return id;
    }
    if (this?.isGNO2GED()) {
      const id = this.getGNO2GEDTreeId();
      if (id !== void 0) return id;
    }
    if (this?.isAhnenblatt()) {
      const id = this.getAhnenblattTreeId();
      if (id !== void 0) return id;
    }
    if (this?.isGeni()) {
      const id = this.getGeniTreeId();
      if (id !== void 0) return id;
    }
    if (this?.isGenoPro()) {
      const id = this.getGenoProTreeId();
      if (id !== void 0) return id;
    }
    return this.getUniversalTreeId();
  }
  getAncestryTreeName() {
    const path = "HEAD.SOUR._TREE.value";
    return get(this, path) || get(this.getGedcom(), path);
  }
  getMyHeritageTreeName() {
    const path = "HEAD.FILE.value";
    const treeDetails = get(this, path) || get(this.getGedcom(), path);
    return treeDetails?.match(
      /Exported by MyHeritage.com from (?<tree>.+) in.+$/
    )?.groups?.tree;
  }
  getFamilySearchTreeId() {
    const rin = get(this, "HEAD.SOUR._TREE.RIN.value") || get(this.getGedcom(), "HEAD.SOUR._TREE.RIN.value");
    if (rin) {
      return rin;
    }
    return "familysearch";
  }
  getFamilySearchTreeName() {
    const treeName = get(this, "HEAD.SOUR._TREE.value") || get(this.getGedcom(), "HEAD.SOUR._TREE.value");
    if (treeName) {
      return treeName;
    }
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    return fileName || "FamilySearch Import";
  }
  getAhnenblattTreeId() {
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    if (fileName) {
      const idMatch = fileName.match(/_(\d+)/);
      if (idMatch) {
        return idMatch[1];
      }
    }
    return void 0;
  }
  getAhnenblattTreeName() {
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    return fileName?.replace(/\.ged$/i, "").replace(/_/g, " ");
  }
  getGeniTreeId() {
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    if (fileName) {
      const idMatch = fileName.match(/_(\d+)/);
      if (idMatch) {
        return idMatch[1];
      }
    }
    return void 0;
  }
  getGeniTreeName() {
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    return fileName?.replace(/\.ged$/i, "").replace(/_/g, " ");
  }
  getGenoProTreeId() {
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    if (fileName) {
      const idMatch = fileName.match(/_(\d+)/);
      if (idMatch) {
        return idMatch[1];
      }
    }
    return void 0;
  }
  getGenoProTreeName() {
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    return fileName?.replace(/\.ged$/i, "").replace(/_/g, " ");
  }
  getGNO2GEDTreeId() {
    const rin = get(this, "HEAD._TREE.RIN.value") || get(this.getGedcom(), "HEAD._TREE.RIN.value");
    if (rin) {
      return rin;
    }
    return `gno_${this._gedcom?.refcount || "unknown"}`;
  }
  getGNO2GEDTreeName() {
    const treeName = get(this, "HEAD._TREE.value") || get(this.getGedcom(), "HEAD._TREE.value");
    if (treeName) {
      return treeName;
    }
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    return fileName || "GNO2GED Export";
  }
  /**
   * Universal tree ID getter for unknown/unrecognized GEDCOM sources
   * Tries to extract an ID from various common locations
   */
  getUniversalTreeId() {
    const sourceType = this.getSourceType();
    const prefix = sourceType ? sourceType.toLowerCase() : "tree";
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    if (fileName) {
      const idMatch = fileName.match(/_(\d+)/);
      if (idMatch) {
        return `${prefix}_${idMatch[1]}`;
      }
    }
    return `${prefix}_${this._gedcom?.refcount || "unknown"}`;
  }
  /**
   * Universal tree name getter for unknown/unrecognized GEDCOM sources
   * Tries to extract a name from various common locations
   */
  getUniversalTreeName() {
    const sourceType = this.getSourceType();
    const prefix = sourceType ? `${sourceType}-` : "";
    const fileName = get(this, "HEAD.FILE.value") || get(this.getGedcom(), "HEAD.FILE.value");
    if (fileName) {
      const cleanName = fileName.replace(/\.ged$/i, "").replace(/_/g, " ");
      return `${prefix}${cleanName}`;
    }
    const sourName = get(this, "HEAD.SOUR.NAME.value") || get(this.getGedcom(), "HEAD.SOUR.NAME.value");
    if (sourName) {
      return `${prefix}${sourName}`;
    }
    const sourValue = get(this, "HEAD.SOUR.value") || get(this.getGedcom(), "HEAD.SOUR.value");
    if (sourValue) {
      return `${prefix}${sourValue}`;
    }
    return `${prefix}Unknown Tree`;
  }
  getTreeName() {
    if (this?.isAncestry()) {
      const name = this.getAncestryTreeName();
      if (name) return name;
    }
    if (this?.isMyHeritage()) {
      const name = this.getMyHeritageTreeName();
      if (name) return name;
    }
    if (this?.isFamilySearch()) {
      const name = this.getFamilySearchTreeName();
      if (name) return name;
    }
    if (this?.isGNO2GED()) {
      const name = this.getGNO2GEDTreeName();
      if (name) return name;
    }
    if (this?.isAhnenblatt()) {
      const name = this.getAhnenblattTreeName();
      if (name) return name;
    }
    if (this?.isGeni()) {
      const name = this.getGeniTreeName();
      if (name) return name;
    }
    if (this?.isGenoPro()) {
      const name = this.getGenoProTreeName();
      if (name) return name;
    }
    return this.getUniversalTreeName();
  }
};
var createProxy = (target) => {
  return new Proxy(target, {
    get(t, prop, receiver) {
      if (prop === "unwrapped") {
        return target;
      }
      if (prop in t) {
        return Reflect.get(t, prop, receiver);
      }
      if (!isOnlyMainProp(prop)) {
        const ref = t.ref;
        if (ref && prop in ref) {
          return Reflect.get(ref, prop, receiver);
        }
      }
    }
  });
};
var createCommon = (gedcom, id, main, parent) => {
  return createProxy(
    new Common(gedcom, id, main, parent)
  );
};
var isOnlyMainProp = (key) => {
  return key === "_gedcom" || key === "_main" || key === "_parent" || key === "_uniqueId" || key === "_type" || key === "_refs" || key === "id" || key === "_id";
};
var isValidKey = (common, key) => {
  const prop = get(common, key);
  return key !== "_gedcom" && key !== "_main" && key !== "_parent" && key !== "_uniqueId" && key !== "_refs" && key !== "_type" && (key === "id" || key === "_id" || key === "value" || key === "_value" || prop instanceof Common || prop instanceof List);
};
var getValidKeys = (common) => {
  return Object.keys(common).filter((key) => {
    return isValidKey(common, key);
  });
};
var getValidTag = (tag) => {
  return tag.replace(/^@@/, "");
};
var getListTag = (tag) => {
  return `@@${tag}`;
};
var isId = (string) => {
  return ID_REG.test(string);
};
var idGetter = (id) => {
  if (id === void 0) {
    return void 0;
  }
  let newId = `${id}`;
  const parts = newId?.match(ID_GETTER_REG)?.groups;
  if (!parts?.at && !parts?.letter) {
    newId = `@U${newId}`;
  } else if (!parts.at) {
    newId = `@${newId}`;
  } else if (!parts.letter) {
    newId = newId.replace(/^@/, "@U");
  }
  if (!newId.endsWith("@")) {
    newId = `${newId}@`;
  }
  return newId;
};

// src/classes/date.ts
var LONG_NOTES = {
  "Abt.": "About",
  "Bef.": "Before",
  "Aft.": "After"
};
var NOTE_MARKER = "####";
var formatDateWithComponents = (date, hasDay, hasMonth, hasYear, noteValue, baseFormat = "dd MMM yyyy", locale) => {
  const hasNote = baseFormat.includes("NOTE");
  let validDateFormat = baseFormat.replace("NOTE", NOTE_MARKER);
  if (!hasDay) {
    validDateFormat = validDateFormat.replace(/d+/g, "");
  }
  if (!hasMonth) {
    validDateFormat = validDateFormat.replace(/[.\-\s/]*M+/g, "");
  }
  if (!hasYear) {
    validDateFormat = validDateFormat.replace(/y+[.\-\s/]*/g, "");
  }
  validDateFormat = validDateFormat.replace(/([.\-\s/])\1+/g, "$1").replace(/^[.\-\s/]+|[.\-\s/]+$/g, "");
  const formattedDate = format(date, validDateFormat, {
    ...locale !== null ? { locale: locale ? locale : getDateLocale() } : {}
  });
  return noteValue && hasNote ? formattedDate.replace(NOTE_MARKER, noteValue).trim() : formattedDate.replace(NOTE_MARKER, "").trim();
};
var CommonDate = class extends Common {
  constructor(gedcom, id, main, parent) {
    super(gedcom, id, main, parent);
    delete this.id;
  }
  set value(value) {
    if (value) {
      const noteRegExp = /^(?<note>[a-zA-Z]+\.)/;
      const noteMatch = value.match(noteRegExp)?.groups;
      let validValue = value;
      if (noteMatch?.note) {
        this.NOTE = this.NOTE || createCommon(this._gedcom, void 0, this.main);
        this.NOTE.value = noteMatch?.note;
        validValue = value.replace(noteRegExp, "");
      }
      const acceptedDate = this.isValidDateFormat(validValue);
      if (acceptedDate) {
        this.DAY = this.DAY || createCommon(this._gedcom, void 0, this.main);
        this.DAY.value = format(acceptedDate, "dd");
        this.MONTH = this.MONTH || createCommon(this._gedcom, void 0, this.main);
        this.MONTH.value = format(acceptedDate, "MMM");
        this.YEAR = this.YEAR || createCommon(this._gedcom, void 0, this.main);
        this.YEAR.value = format(acceptedDate, "yyyy");
        this._date = acceptedDate;
        this._value = value;
      } else {
        let fixedValue = validValue;
        if (/\d{4} [A-Za-z]+\s*$/.test(validValue)) {
          fixedValue = `${validValue} 1`;
        } else if (/^\s*[A-Za-z]+ \d{4}/.test(validValue)) {
          fixedValue = `1 ${validValue}`;
        }
        this._date = new Date(fixedValue);
        this._value = value;
        if (this._date && isValid(this._date)) {
          const yearMonthDay = /[\dA-Za-z]+ [\dA-Za-z]+ [\dA-Za-z]+/.test(validValue);
          const yearMonth = /[\dA-Za-z]+ [\dA-Za-z]+/.test(
            validValue
          );
          const year = /[\dA-Za-z]+/.test(validValue);
          if (yearMonthDay) {
            this.DAY = this.DAY || createCommon(this._gedcom, void 0, this.main);
            this.DAY.value = format(this._date, "dd");
          }
          if (yearMonth || yearMonthDay) {
            this.MONTH = this.MONTH || createCommon(this._gedcom, void 0, this.main);
            this.MONTH.value = format(this._date, "MMM");
          }
          if (year || yearMonth || yearMonthDay) {
            this.YEAR = this.YEAR || createCommon(this._gedcom, void 0, this.main);
            this.YEAR.value = format(this._date, "yyyy");
          }
        }
      }
    }
  }
  get value() {
    const hasDay = !!this.DAY?.value;
    const hasMonth = !!this.MONTH?.value;
    const hasYear = !!this.YEAR?.value;
    if (!this._date || !isValid(this._date) || !hasDay && !hasMonth && !hasYear) {
      return this._value;
    }
    return formatDateWithComponents(
      this._date,
      hasDay,
      hasMonth,
      hasYear,
      this.NOTE?.value
    );
  }
  get rawValue() {
    return this._date;
  }
  assign(name, value, unique = false) {
    if (!["DAY", "MONTH", "YEAR"].includes(name)) {
      return super.assign(name, value, unique);
    } else {
      this.set(name, value);
    }
    return this.get(name);
  }
  isValidDateFormat(value) {
    let validDate;
    ACCEPTED_DATE_FORMATS.find((acceptedFormat) => {
      const date = parse(value, acceptedFormat, /* @__PURE__ */ new Date());
      if (isValid(date)) {
        validDate = date;
        return true;
      }
      return false;
    });
    return validDate;
  }
  toNote(short = true) {
    const note = this.NOTE?.value?.trim();
    if (!short) {
      return LONG_NOTES[note];
    }
    return note;
  }
  toValue(dateFormat = "dd MMM yyyy", locale) {
    const hasDay = !!this.DAY?.value;
    const hasMonth = !!this.MONTH?.value;
    const hasYear = !!this.YEAR?.value;
    if (!this._date || !isValid(this._date) || !hasDay && !hasMonth && !hasYear) {
      return this._value;
    }
    return formatDateWithComponents(
      this._date,
      hasDay,
      hasMonth,
      hasYear,
      this.NOTE?.value,
      dateFormat,
      locale
    );
  }
  exportValue() {
    return this.toValue("NOTE dd MMM yyyy", null);
  }
  inRange(range, trueIfNoYear = false) {
    return inRange(this.YEAR?.value, range, trueIfNoYear);
  }
};
var createCommonDate = (gedcom, id, main, parent) => {
  return createProxy(
    new CommonDate(gedcom, id, main, parent)
  );
};
var isCommonDate = (value) => {
  return !!value && value !== null && typeof value === "object" && "_date" in value;
};

// src/utils/get-family-with.ts
var getFamilyWith = (person1, person2, famType = "FAMS") => {
  return person1?.get(famType)?.toValueList().find((fam) => {
    const family = fam;
    return famType === "FAMC" || family?.HUSB?.value === person2 || family?.WIFE?.value === person2;
  });
};

// src/constants/orders.ts
var BIRTH_ASC = {
  "BIRT.DATE": {
    direction: "ASC",
    getter: (value, raw) => {
      if (isCommonDate(raw) && raw.rawValue) {
        return raw.rawValue.getTime();
      }
      return new Date(value).getTime();
    }
  }
};
var BIRTH_DESC = {
  "BIRT.DATE": {
    direction: "DESC",
    getter: (value, raw) => {
      if (isCommonDate(raw) && raw.rawValue) {
        return raw.rawValue.getTime();
      }
      return new Date(value).getTime();
    }
  }
};
var DEATH_ASC = {
  "DEAT.DATE": {
    direction: "ASC",
    getter: (value, raw) => {
      if (isCommonDate(raw) && raw.rawValue) {
        return raw.rawValue.getTime();
      }
      return new Date(value).getTime();
    }
  }
};
var DEATH_DESC = {
  "DEAT.DATE": {
    direction: "DESC",
    getter: (value, raw) => {
      if (isCommonDate(raw) && raw.rawValue) {
        return raw.rawValue.getTime();
      }
      return new Date(value).getTime();
    }
  }
};
var DATE_ASC = {
  DATE: {
    direction: "ASC",
    getter: (value, raw) => {
      if (isCommonDate(raw) && raw.rawValue) {
        return raw.rawValue.getTime();
      }
      return new Date(value).getTime();
    }
  }
};
var chars = [
  "a",
  "\xE0",
  "\xE1",
  "\xE2",
  "\xE4",
  "\xE6",
  "\xE3",
  "\xE5",
  "\u0101",
  "b",
  "c",
  "\xE7",
  "\u0107",
  "\u010D",
  "cs",
  "d",
  "dz",
  "dzs",
  "e",
  "\xE8",
  "\xE9",
  "\xEA",
  "\xEB",
  "\u0113",
  "\u0117",
  "\u0119",
  "f",
  "g",
  "gy",
  "h",
  "i",
  "\xEE",
  "\xEF",
  "\xED",
  "\u012B",
  "\u012F",
  "\xEC",
  "j",
  "k",
  "l",
  "\u0142",
  "ly",
  "m",
  "n",
  "\xF1",
  "\u0144",
  "ny",
  "o",
  "\xF3",
  "\xF6",
  "\u0151",
  "\xF4",
  "\xF2",
  "\xF5",
  "\u0153",
  "\xF8",
  "\u014D",
  "p",
  "q",
  "r",
  "s",
  "\u015B",
  "\u0161",
  "\xDF",
  "sz",
  "t",
  "ty",
  "u",
  "\xFA",
  "\xFC",
  "\u0171",
  "\xFB",
  "\xF9",
  "\u016B",
  "v",
  "w",
  "x",
  "y",
  "z",
  "zs"
];
chars.reduce((acc, curr, idx) => {
  acc[curr] = idx;
  return acc;
}, {});
var getBirthAsc = (itemA, _keyA, itemB, _keyB) => {
  const birthA = itemA?.BIRT?.index(0);
  const birthB = itemB?.BIRT?.index(0);
  const valueA = birthA?.DATE?.rawValue;
  const valueB = birthB?.DATE?.rawValue;
  if (valueA?.getTime() === valueB?.getTime()) {
    return 0;
  }
  if (!valueA || valueB && valueA > valueB) {
    return 1;
  }
  return -1;
};
var getMarriageAsc = (person, famType = "FAMS") => (itemA, _keyA, itemB, _keyB) => {
  if (!person && famType === "FAMS") {
    return 0;
  }
  const familyA = getFamilyWith(itemA, person);
  const familyB = getFamilyWith(itemB, person);
  const marriageA = familyA?.MARR?.DATE?.rawValue;
  const marriageB = familyB?.MARR?.DATE?.rawValue;
  if (marriageA?.getTime() === marriageB?.getTime()) {
    return 0;
  }
  if (!marriageA || marriageB && marriageA > marriageB) {
    return 1;
  }
  return -1;
};
var getMarriageAscAndChildBirth = (person) => (itemA, keyA, itemB, keyB) => {
  const marriageSort = getMarriageAsc(person)(itemA, keyA, itemB, keyB);
  if (marriageSort !== 0) {
    return marriageSort;
  }
  const familyA = getFamilyWith(itemA, person);
  const familyB = getFamilyWith(itemB, person);
  const childA = familyA?.getChildren().orderBy(BIRTH_ASC).index(0);
  const childB = familyB?.getChildren().orderBy(BIRTH_ASC).index(0);
  return getBirthAsc(childA, keyA, childB);
};
var getGedcomId = (gedcom) => {
  if (!gedcom) {
    return "unknown";
  }
  const treeId = gedcom.getTreeId?.() || "";
  const treeName = gedcom.getTreeName?.() || "";
  const sanitizedName = treeName.toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "");
  if (treeId && sanitizedName) {
    return `${treeId}_${sanitizedName}`;
  } else if (treeId) {
    return treeId;
  } else if (sanitizedName) {
    return sanitizedName;
  }
  return `gedcom_${gedcom.refcount}`;
};
var caches = {
  pathCache: {},
  relativesOnDegreeCache: {},
  relativesOnLevelCache: {},
  profilePictureCache: {}
};
var cacheDbs;
var getCacheDbs = () => {
  if (!cacheDbs) {
    const getInstance = getCacheManagerFactory();
    cacheDbs = {
      pathCache: getInstance(
        "ftv",
        "Main",
        "path",
        true
      ),
      relativesOnDegreeCache: getInstance("ftv", "Main", "path", true),
      relativesOnLevelCache: getInstance(
        "ftv",
        "Main",
        "path",
        true
      ),
      profilePictureCache: getInstance(
        "ftv",
        "Main",
        "images",
        false
      )
    };
  }
  return cacheDbs;
};
var storeCache = {
  // NOTE: pathCache, relativesOnLevelCache, and relativesOnDegreeCache are intentionally
  // kept in memory only. These debounced functions exist to satisfy the type system
  // but are never called.
  pathCache: debounce((value) => {
    if (value) {
      getCacheDbs().pathCache.setItem(value);
    }
  }, 50),
  relativesOnLevelCache: debounce((value) => {
    if (value) {
      getCacheDbs().relativesOnLevelCache.setItem(value);
    }
  }, 50),
  relativesOnDegreeCache: debounce((value) => {
    if (value) {
      getCacheDbs().relativesOnDegreeCache.setItem(value);
    }
  }, 50),
  // profilePictureCache IS persisted to IndexedDB
  profilePictureCache: debounce((value) => {
    if (value) {
      getCacheDbs().profilePictureCache.setItem(value);
    }
  }, 100)
};
var relativesCache = (cacheKey) => (gedcom, key, subKey, value) => {
  const gedcomId = getGedcomId(gedcom);
  const fullKey = `${gedcomId}:${key}`;
  const cache = caches[cacheKey];
  if (!cache) {
    caches[cacheKey] = {};
  }
  if (value) {
    const typedCache2 = caches[cacheKey];
    if (!typedCache2[fullKey]) {
      typedCache2[fullKey] = {};
    }
    typedCache2[fullKey][subKey] = value;
    return typedCache2[fullKey][subKey];
  }
  const typedCache = caches[cacheKey];
  return typedCache?.[fullKey]?.[subKey];
};
var pathCache = (gedcom, key, value) => {
  const gedcomId = getGedcomId(gedcom);
  const fullKey = `${gedcomId}:${key}`;
  if (!caches.pathCache) {
    caches.pathCache = {};
  }
  if (value && caches.pathCache) {
    caches.pathCache[fullKey] = value;
    return caches.pathCache[fullKey];
  }
  return caches.pathCache?.[fullKey];
};
var profilePictureCache = (gedcom, key, value) => {
  const gedcomId = getGedcomId(gedcom);
  const fullKey = `${gedcomId}:${key}`;
  if (!caches.profilePictureCache) {
    caches.profilePictureCache = {};
  }
  if (value && caches.profilePictureCache) {
    caches.profilePictureCache[fullKey] = value;
    storeCache.profilePictureCache(caches.profilePictureCache);
    return caches.profilePictureCache[fullKey];
  }
  const cached = caches.profilePictureCache?.[fullKey];
  return cached;
};

// src/utils/get-places.ts
var getPlaces = (common, type = ["ALL" /* All */], maxLevel = 1, level = 0, mainKey) => {
  const types = Array.isArray(type) ? type : [type];
  const places = [];
  if (!common?.toList || level > maxLevel) {
    return places;
  }
  const commonList = common.toList();
  commonList.forEach((item, _, index) => {
    Object.entries(item).forEach(
      ([key, value]) => {
        if (!/^[_A-Z0-9]+$/.test(key)) {
          return;
        }
        if (level === 0 && !types.includes("ALL" /* All */) && (!types.includes(key) && [
          "BIRT" /* Birth */,
          "MARR" /* Marriage */,
          "DEAT" /* Death */
        ].includes(key) || !types.includes("EVEN" /* Events */) && ![
          "BIRT" /* Birth */,
          "MARR" /* Marriage */,
          "DEAT" /* Death */,
          "_MILT" /* Military */,
          "_MILTID" /* MilitaryId */
        ].includes(key) || !types.includes("_MILT" /* Military */) && !types.includes("_MILTID" /* MilitaryId */) && ![
          "BIRT" /* Birth */,
          "MARR" /* Marriage */,
          "DEAT" /* Death */,
          "EVEN" /* Events */
        ].includes(key))) {
          return;
        }
        if (key === "PLAC") {
          value.toList().forEach((place) => {
            places.push({
              index,
              place: place.toValue(),
              obj: place,
              key: mainKey || key
            });
          });
        } else {
          getPlaces(
            value,
            types,
            maxLevel,
            level + 1,
            mainKey || key
          ).forEach((place) => {
            const usedValue = value instanceof List ? value.index(place.index) : value;
            places.push({
              place: place.place,
              obj: usedValue,
              ref: place.obj,
              index: place.index,
              key: place.key
            });
          });
        }
      }
    );
  });
  return places;
};
var implemented = (type, ...args) => {
};

// src/utils/media-utils.ts
var getFileExtension = (filename) => {
  const match = filename.match(/\.([^.]+)$/);
  return match ? match[1] : "";
};
var isImageFormat = (format2) => {
  if (!format2) return false;
  const imageFormats = [
    "jpg",
    "jpeg",
    "png",
    "gif",
    "bmp",
    "webp",
    "svg",
    "tiff",
    "tif"
  ];
  return imageFormats.includes(format2.toLowerCase());
};

// src/factories/place-parser-provider.ts
var placeParserProvider;
var setPlaceParserProvider = (parser) => {
  placeParserProvider = parser;
};
var getPlaceParserProvider = () => {
  return placeParserProvider;
};
var resetPlaceParserProvider = () => {
  placeParserProvider = void 0;
};

// src/factories/place-translator-provider.ts
var placeTranslatorProvider;
var setPlaceTranslatorProvider = (translator) => {
  placeTranslatorProvider = translator;
};
var getPlaceTranslatorProvider = () => {
  return placeTranslatorProvider;
};
var resetPlaceTranslatorProvider = () => {
  placeTranslatorProvider = void 0;
};
var uniqueItemsCache = /* @__PURE__ */ new WeakMap();
var setNestedGroup = (obj, key, value, uniqueCounting = true) => {
  const parts = Array.isArray(key) ? key : key.split(/,\s*/);
  set(obj, [...parts, "items"], value);
  parts.forEach((_, index) => {
    if (!uniqueCounting) {
      const key2 = [...parts.toSpliced(index + 1), "length"];
      const length = get(obj, key2);
      set(obj, key2, (length ?? 0) + value.length);
      return;
    }
    const pathKey = [...parts.toSpliced(index + 1)];
    const node = get(obj, pathKey);
    if (node) {
      let uniqueItems = uniqueItemsCache.get(node);
      if (!uniqueItems) {
        uniqueItems = new Individuals();
        uniqueItemsCache.set(node, uniqueItems);
      }
      uniqueItems.merge(value);
      set(obj, [...pathKey, "length"], uniqueItems.length);
    }
  });
};

// src/utils/place-parser.ts
var getPlaceParts = (place = []) => {
  const partsString = Array.isArray(place) ? place.filter(Boolean).join(", ") : place;
  if (!partsString) {
    return [{}];
  }
  const parts = partsString.split(/,\s*/).map((p) => p.trim()).filter(Boolean);
  if (parts.length === 0) {
    return [{}];
  }
  const result = {};
  if (parts.length === 1) {
    result.town = parts[0];
  } else if (parts.length === 2) {
    result.town = parts[0];
    result.country = parts[1];
  } else if (parts.length === 3) {
    result.town = parts[0];
    result.county = parts[1];
    result.country = parts[2];
  } else {
    result.leftParts = parts.slice(0, parts.length - 3);
    result.town = parts[parts.length - 3];
    result.county = parts[parts.length - 2];
    result.country = parts[parts.length - 1];
  }
  return [result];
};

// src/utils/place-translator.ts
var placeTranslator = (place) => {
  return Array.isArray(place) ? place.join(", ") : place;
};

// src/classes/indis.ts
var Individuals = class _Individuals extends List {
  copy() {
    return super.copy(_Individuals);
  }
  except(item) {
    return super.except(item, _Individuals);
  }
  filter(filters) {
    return super.filter(filters, _Individuals);
  }
  find(filters) {
    return super.find(filters, _Individuals);
  }
  unattachedFilter(useUnattached) {
    if (useUnattached) {
      return this;
    }
    return this.filter((item) => !item.isUnattachedMember());
  }
  orderBy(orders) {
    return super.orderBy(orders, _Individuals);
  }
  getFirstEvent() {
    const firstBirthPerson = this.getFirstBirth();
    const firstDeathPerson = this.getFirstDeath();
    const firstBirth = firstBirthPerson?.BIRT?.index(0);
    const firstDeath = firstDeathPerson?.DEAT?.index(0);
    if (!firstBirth?.DATE?.rawValue && !firstDeath?.DATE?.rawValue) {
      return void 0;
    }
    if (!firstBirth?.DATE?.rawValue) {
      return firstDeathPerson;
    }
    if (!firstDeath?.DATE?.rawValue) {
      return firstBirthPerson;
    }
    if (firstDeath.DATE.rawValue < firstBirth.DATE.rawValue) {
      return firstDeathPerson;
    }
    return firstBirthPerson;
  }
  getLastEvent() {
    const lastDeathPerson = this.getLastDeath();
    const lastBirthPerson = this.getLastBirth();
    const lastBirth = lastBirthPerson?.BIRT?.index(0);
    const lastDeath = lastDeathPerson?.DEAT?.index(0);
    if (!lastBirth?.DATE?.rawValue && !lastDeath?.DATE?.rawValue) {
      return void 0;
    }
    if (!lastDeath?.DATE?.rawValue) {
      return lastBirthPerson;
    }
    if (!lastBirth?.DATE?.rawValue) {
      return lastDeathPerson;
    }
    if (lastDeath.DATE.rawValue < lastBirth.DATE.rawValue) {
      return lastBirthPerson;
    }
    return lastDeathPerson;
  }
  getFirstBirth() {
    return this.filter((item) => {
      const birth = item.BIRT?.index(0);
      if (birth?.DATE?.rawValue?.getTime() === 0) {
        return true;
      }
      return !!birth?.DATE?.rawValue?.getTime();
    }).orderBy(BIRTH_ASC).index(0);
  }
  getLastBirth() {
    return this.filter((item) => {
      const birth = item.BIRT?.index(0);
      if (birth?.DATE?.rawValue?.getTime() === 0) {
        return true;
      }
      return !!birth?.DATE?.rawValue?.getTime();
    }).orderBy(BIRTH_DESC).index(0);
  }
  getFirstDeath() {
    return this.filter((item) => {
      const death = item.DEAT?.index(0);
      if (death?.DATE?.rawValue?.getTime() === 0) {
        return true;
      }
      return !!death?.DATE?.rawValue?.getTime();
    }).orderBy(DEATH_ASC).index(0);
  }
  getLastDeath() {
    return this.filter((item) => {
      const death = item.DEAT?.index(0);
      if (death?.DATE?.rawValue?.getTime() === 0) {
        return true;
      }
      return !!death?.DATE?.rawValue?.getTime();
    }).orderBy(DEATH_DESC).index(0);
  }
  groupBy(group, sort, minOccurance = -1, lessOccuranceLabel) {
    return super.groupBy(
      group,
      sort,
      minOccurance,
      lessOccuranceLabel,
      _Individuals
    );
  }
  groupByFirstLetters(sort = "group", minOccurance = -1, settings) {
    return this.groupBy(
      (item) => {
        const { letter } = nameFormatter(item, settings);
        return letter.trim() || "Unknown";
      },
      sort,
      minOccurance,
      "Letters appearing less than 5 times"
    );
  }
  groupBySurnames(sort, minOccurance = -1) {
    return this.groupBy(
      "NAME.SURN",
      sort,
      minOccurance,
      "Surnames appearing less than 5 times"
    );
  }
  groupByTimeRanges(range, sort = "group", minOccurance = -1) {
    return this.groupBy(
      {
        ["BIRT.DATE.YEAR"]: {
          getter: (year) => {
            if (!Number(year)) {
              return;
            }
            if (range === 1 /* Year */) {
              return `${Math.floor(Number(year) / range) * range}`;
            }
            return `${Math.floor(Number(year) / range) * range + 1}-${Math.ceil((Number(year) + 1) / range) * range}`;
          }
        }
      },
      sort,
      minOccurance,
      "Time ranges appearing less than 5 times"
    );
  }
  nestedGroupByPlacesAndTimeRanges(sort, minOccurance = -1, uniqueCounting = true) {
    const placesWithYear = this.groupBy(
      (item) => item.getPlaces().map((place) => {
        const date = place.obj?.get("DATE");
        if (date?.YEAR) {
          return `${place.place}, ${date.YEAR}`;
        }
        return `${place.place}, Unknown`;
      }).filter(Boolean),
      sort,
      minOccurance,
      "Places appearing less than 5 times"
    );
    return Object.entries(placesWithYear).reduce(
      (acc, [place, indis]) => {
        setNestedGroup(acc, place, indis, uniqueCounting);
        return acc;
      },
      {}
    );
  }
  nestedGroupByTimeRangesAndPlaces(sort, minOccurance = -1, uniqueCounting = true) {
    const placesWithYear = this.groupBy(
      (item) => item.getPlaces().map((place) => {
        const date = place.obj?.get("DATE");
        if (date?.YEAR) {
          return `${date.YEAR}, ${place.place}`;
        }
        return `Unknown, ${place.place}`;
      }).filter(Boolean),
      sort,
      minOccurance,
      "Places appearing less than 5 times"
    );
    return Object.entries(placesWithYear).reduce(
      (acc, [place, indis]) => {
        setNestedGroup(acc, place, indis, uniqueCounting);
        return acc;
      },
      {}
    );
  }
  nestedGroupByPlaces(sort, minOccurance = -1, partOrder = "reverse", translate = true, uniqueCounting = true, groupKeyTranslator) {
    const places = this.groupByPlaces(
      sort,
      minOccurance,
      partOrder,
      translate,
      groupKeyTranslator
    );
    return Object.entries(places).reduce(
      (acc, [place, indis]) => {
        setNestedGroup(acc, place, indis, uniqueCounting);
        return acc;
      },
      {}
    );
  }
  groupByPlaces(sort, minOccurance = -1, partOrder = "reverse", translate = true, groupKeyTranslator) {
    return this.groupBy(
      (item) => item.getPlaces().map((place) => {
        const customPlaceParser = getPlaceParserProvider();
        const placeParts = customPlaceParser ? customPlaceParser(place.place || "") : getPlaceParts(place.place || "");
        const {
          leftParts = [],
          town,
          county,
          country
        } = placeParts[0] || {};
        const normalizedParts = [
          ...leftParts,
          town,
          county,
          country
        ].filter(Boolean);
        const orderedParts = partOrder === "original" ? normalizedParts : normalizedParts.toReversed();
        let newPlace = "";
        if (groupKeyTranslator) {
          newPlace = groupKeyTranslator(orderedParts);
        } else if (translate) {
          const customPlaceTranslator = getPlaceTranslatorProvider();
          if (customPlaceTranslator) {
            newPlace = customPlaceTranslator(orderedParts);
          } else {
            newPlace = placeTranslator(orderedParts);
          }
        } else {
          newPlace = orderedParts.join(", ");
        }
        if (!newPlace) {
          return void 0;
        }
        return {
          marker: place.place,
          group: newPlace
        };
      }).filter(Boolean),
      sort,
      minOccurance,
      "Places appearing less than 5 times"
    );
  }
  getFacts() {
    const facts = new List();
    this.values().forEach((indi) => {
      const indiFacts = indi?.getFacts();
      if (indi?.id && indiFacts) {
        facts.merge(indiFacts);
      }
    });
    return facts;
  }
  isRelativeOf(type, indi, every = false) {
    const usedIndi = typeof indi === "string" ? this.index(0)?.getGedcom()?.indi(indi) : indi;
    let getter;
    if (type === "sibling") {
      getter = "getSiblings";
    }
    if (type === "parent") {
      getter = "getParents";
    }
    if (type === "child") {
      getter = "getChildren";
    }
    if (type === "spouse") {
      getter = "getCoParents";
    }
    if (type === "parentInLaw") {
      getter = "getParentsInLaw";
    }
    if (type === "childInLaw") {
      getter = "getChildrenInLaw";
    }
    if (!usedIndi || !getter) {
      return false;
    }
    const relatives = usedIndi[getter]();
    const thisIds = this.keys();
    if (relatives.length <= 0 || this.length <= 0) {
      return false;
    }
    const filtered = relatives.filter({ id: thisIds });
    if (every) {
      return filtered.length === this.length ? filtered.index(0)?.id || true : false;
    }
    return filtered.length > 0 ? filtered.index(0)?.id || true : false;
  }
  isSiblingOf(indi, every = false) {
    return this.isRelativeOf("sibling", indi, every);
  }
  isSpouseOf(indi, every = false) {
    return this.isRelativeOf("spouse", indi, every);
  }
  isParentOf(indi, every = false) {
    return this.isRelativeOf("parent", indi, every);
  }
  isChildOf(indi, every = false) {
    return this.isRelativeOf("child", indi, every);
  }
  isParentInLawOf(indi, every = false) {
    return this.isRelativeOf("parentInLaw", indi, every);
  }
  isChildInLawOf(indi, every = false) {
    return this.isRelativeOf("childInLaw", indi, every);
  }
  splitByFamily(type, relativeTo) {
    const splittedList = {};
    const familiesRelativeTo = relativeTo?.get(type !== "Parents" ? "FAMS" : "FAMC")?.toValueList().keys() ?? [];
    let lengthOfIndividuals = 0;
    this.forEach((indi) => {
      const families = indi.get(type === "Spouses" ? "FAMS" : "FAMC")?.toValueList();
      if (families) {
        families.keys().forEach((fKey) => {
          const famKey = fKey;
          if (familiesRelativeTo && !familiesRelativeTo.includes(famKey)) {
            return;
          }
          if (!splittedList[famKey]) {
            splittedList[famKey] = new _Individuals();
          }
          splittedList[famKey]?.append(indi);
          lengthOfIndividuals++;
        });
      }
    });
    if (type === "Spouses" && relativeTo) {
      const missingFams = difference(
        familiesRelativeTo,
        Object.keys(splittedList)
      );
      if (missingFams.length) {
        missingFams.forEach((fKey) => {
          const famKey = fKey;
          if (familiesRelativeTo && !familiesRelativeTo.includes(famKey)) {
            return;
          }
          const missingChildren = relativeTo.getGedcom()?.fam(famKey)?.getChildren();
          if (!splittedList[famKey] && missingChildren?.length) {
            splittedList[famKey] = new _Individuals();
          }
        });
      }
    }
    return {
      items: splittedList,
      lengthOfFamily: Object.keys(splittedList).length,
      lengthOfIndividuals
    };
  }
  getRelativesOnDegree(degree = 0) {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getRelativesOnLevel(degree));
      }
    });
    return persons;
  }
  getRelativesOnLevel(level = 0, filter) {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getRelativesOnLevel(level, filter));
      }
    });
    return persons;
  }
  getAscendants(level = 0, filter) {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getAscendants(level, filter));
      }
    });
    return persons;
  }
  getDescendants(level = 0, filter) {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getDescendants(level, filter));
      }
    });
    return persons;
  }
  getAllAscendants(individuals) {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getAllAscendants(individuals));
      }
    });
    return persons;
  }
  getAllDescendants(individuals, containDescendantsInLaw = false) {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(
          indi.getAllDescendants(individuals, containDescendantsInLaw)
        );
      }
    });
    return persons;
  }
  getSiblings() {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getSiblings());
      }
    });
    return persons;
  }
  getChildren() {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getChildren());
      }
    });
    return persons;
  }
  getParents() {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getParents());
      }
    });
    return persons;
  }
  getCoParents() {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getCoParents());
      }
    });
    return persons;
  }
  getSpouses() {
    const persons = new _Individuals();
    this.values().forEach((indi) => {
      if (indi) {
        persons.merge(indi.getCoParents());
      }
    });
    return persons;
  }
  toName() {
    return super.toProp("NAME")?.toValue();
  }
  toList() {
    return new _Individuals().concat(this.getItems());
  }
};

// src/classes/fam.ts
var Fam = class extends Common {
  getFamilyMembers(type) {
    const familyMembers = new Individuals();
    this.get(type)?.toList().forEach((item) => {
      const indiId = item.value;
      const indi = this._gedcom?.indi(indiId);
      if (indi) {
        familyMembers.item(indiId, indi);
      }
    });
    return familyMembers;
  }
  getChildren() {
    return this.getFamilyMembers("CHIL");
  }
  getHusband() {
    return this.getFamilyMembers("HUSB");
  }
  getWife() {
    return this.getFamilyMembers("WIFE");
  }
  getParents() {
    return this.getHusband().copy().merge(this.getWife());
  }
  toList() {
    return new Families().concat(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.id ? { [this.id]: this } : { ...[this] }
    );
  }
};

// src/classes/fams.ts
var Families = class _Families extends List {
  copy() {
    return super.copy(_Families);
  }
  except(item) {
    return super.except(item, _Families);
  }
  filter(filters) {
    return super.filter(filters, _Families);
  }
  find(filters) {
    return super.find(filters, _Families);
  }
  orderBy(orders) {
    return super.orderBy(orders, _Families);
  }
  getParents() {
    const persons = new Individuals();
    this.values().forEach((fam) => {
      if (fam && fam instanceof Fam) {
        persons.merge(fam.getParents());
      }
    });
    return persons;
  }
  getChildren() {
    const persons = new Individuals();
    this.values().forEach((fam) => {
      if (fam && fam instanceof Fam) {
        persons.merge(fam.getChildren());
      }
    });
    return persons;
  }
  toList() {
    return new _Families().concat(this.getItems());
  }
};

// src/classes/name.ts
var CommonName = class extends Common {
  constructor(gedcom, id, main, parent) {
    super(gedcom, id, main, parent);
    delete this.id;
  }
  set value(value) {
    if (value && this._main instanceof Indi) {
      const nameParts = value.match(
        /(?<givenname>[^/]*)(\/(?<surname>[^/]*)\/)?(?<suffix>.*)$/
      )?.groups;
      if (nameParts?.givenname?.trim()) {
        this.GIVN = this.GIVN || createCommon(this._gedcom, void 0, this.main);
        this.GIVN.isListable = false;
        this.GIVN.value = nameParts.givenname.trim();
      }
      if (nameParts?.surname?.trim()) {
        this.SURN = this.SURN || createCommon(this._gedcom, void 0, this.main);
        this.SURN.value = nameParts.surname.trim();
        this.SURN.isListable = false;
      }
      if (nameParts?.suffix?.trim()) {
        this.NSFX = this.NSFX || createCommon(this._gedcom, void 0, this.main);
        this.NSFX.value = nameParts.suffix.trim();
        this.NSFX.isListable = false;
      }
    }
    this._value = value;
  }
  get value() {
    if (this._value === void 0 && this.DISPLAY?.value) {
      return this.DISPLAY?.value;
    }
    return this._value;
  }
  name(value) {
    this.value = value;
    return this;
  }
};
var createCommonName = (gedcom, id, main, parent) => {
  return createProxy(
    new CommonName(gedcom, id, main, parent)
  );
};

// src/classes/indi.ts
var ALLOWED_FACTS = [
  "EVEN",
  "OCCU",
  "OCCUPATIONS",
  "RESI",
  "BAPT",
  "CHRI",
  "CHR",
  "BURI",
  "EDUC",
  "GRAD",
  "DIV",
  "IMMI",
  "RELI",
  "_MILT",
  "_MILTID",
  "FACT",
  "_ORIG"
];
var DISALLOWED_CUSTOM_FACTS = ["DNA Test", "Newspaper"];
var CustomFactRenderers = {
  AKA: (label, fact, indi) => {
    const originalNameObj = indi?.NAME;
    const note = fact.get("NOTE");
    const factName = note?.toValue();
    const name = createCommonName(fact.getGedcom(), void 0, indi);
    const suffix = originalNameObj?.NSFX?.toValue();
    const surname = originalNameObj?.SURN?.toValue();
    const givenname = originalNameObj?.GIVN?.toValue();
    const nameParts = [
      ...givenname?.split(" ").map((s) => ({ name: s, givenname: true })) ?? [],
      ...surname?.split(" ").map((s) => ({ name: s, surname: true })) ?? [],
      ...suffix?.split(" ").map((s) => ({ name: s, suffix: true })) ?? []
    ].filter(Boolean);
    const factParts = factName?.split(" ");
    if (nameParts.length === factParts?.length && !(note instanceof CommonName)) {
      const guessedSuffix = [];
      const guessedSurname = [];
      const guessedGivenname = [];
      factParts.forEach((f, i) => {
        const namePart = nameParts[i];
        if (!f) {
          return;
        }
        if (namePart.suffix) {
          guessedSuffix.push(f);
        } else if (namePart.givenname) {
          guessedGivenname.push(f);
        } else if (namePart.surname) {
          guessedSurname.push(f);
        }
      });
      const allNamePart = [];
      if (guessedGivenname.length) {
        allNamePart.push(guessedGivenname.join(" "));
      }
      if (guessedSurname.length) {
        allNamePart.push(`/${guessedSurname.join(" ")}/`);
      }
      if (guessedSuffix.length) {
        allNamePart.push(guessedSuffix.join(" "));
      }
      name.value = allNamePart.join(" ");
    } else {
      name.value = factName;
    }
    fact.set("NOTE", name);
    label.value = "AKA";
    fact.set("_LABEL", label);
  }
};
var relativesOnLevelCache = relativesCache("relativesOnLevelCache");
var relativesOnDegreeCache = relativesCache("relativesOnDegreeCache");
var Indi = class extends Common {
  toName() {
    return this.get("NAME")?.toValue();
  }
  toNaturalName() {
    return this.get("NAME")?.toValue()?.replaceAll("/", "");
  }
  toList() {
    return new Individuals().concat(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      this.id ? { [this.id]: this } : { ...[this] }
    );
  }
  generateTree(props) {
    const {
      generations,
      level,
      index,
      showDescendants,
      offspringSpouses,
      ancestorSpouses,
      onlyDescendants,
      maxLevel,
      minLevel,
      drawNonBiologicalAncestors,
      ancestorSiblings
    } = props;
    if (!this.id) {
      throw new Error("Indi must have an id");
    }
    const isMainDescendantLevel = showDescendants && level <= 1;
    const isDescendantLevel = showDescendants && level < 1;
    if (level < 1 && !isDescendantLevel) {
      throw new Error("Arguments 2 must be greater than 0.");
    }
    if (maxLevel !== void 0 && maxLevel < level || minLevel !== void 0 && minLevel > level) {
      return this;
    }
    const gens = generations ?? {
      existed: {},
      tree: {},
      halves: {},
      lastItems: {}
    };
    const families = this.getFamiliesBiologicalFirst("FAMC");
    const familiesAllMember = families.getParents();
    families.forEach((fam, famKey, famIndex) => {
      if (!this.id) {
        return;
      }
      const id = fam.id;
      let father = isDescendantLevel ? void 0 : fam.getHusband().index(0);
      let mother = isDescendantLevel ? void 0 : fam.getWife().index(0);
      const children = isMainDescendantLevel && famIndex > 0 ? this.getChildren() : void 0;
      if (father?.id && (gens.existed[father.id] === "yes" /* YES */ || !drawNonBiologicalAncestors && this.getParentType(father.id) !== "biological" /* BIOLOGICAL */)) {
        father = void 0;
      }
      if (mother?.id && (gens.existed[mother.id] === "yes" /* YES */ || !drawNonBiologicalAncestors && this.getParentType(mother.id) !== "biological" /* BIOLOGICAL */)) {
        mother = void 0;
      }
      if (!id && !father && !mother && (!children?.length || famIndex > 0)) {
        return this;
      }
      let fatherSpouses;
      if (father && ancestorSpouses) {
        fatherSpouses = father.getCoParents().orderBy(getMarriageAscAndChildBirth(father.id));
        if (drawNonBiologicalAncestors) {
          fatherSpouses = fatherSpouses.copy().exclude(familiesAllMember);
        }
        if (mother) {
          fatherSpouses = fatherSpouses.except(mother);
        }
      }
      let motherSpouses;
      if (mother && ancestorSpouses) {
        motherSpouses = mother.getCoParents().orderBy(getMarriageAscAndChildBirth(mother.id));
        if (drawNonBiologicalAncestors) {
          motherSpouses = motherSpouses.copy().exclude(familiesAllMember);
        }
        if (father) {
          motherSpouses = motherSpouses.except(father);
        }
      }
      let childIndex = 0;
      const amount = Math.pow(2, level - 1);
      const half = Math.floor(amount / 2);
      const closers = {
        id: id || this.id,
        index,
        father: isDescendantLevel ? this.isMale() ? this : void 0 : onlyDescendants ? void 0 : father?.generateTree({
          generations: gens,
          level: level + 1,
          famIndex: famIndex * 2,
          index: index * 2,
          ancestorSpouses,
          ancestorSiblings,
          maxLevel,
          minLevel,
          drawNonBiologicalAncestors
        }),
        fatherSpouses: fatherSpouses && (fatherSpouses.values() ?? []),
        mother: isDescendantLevel ? this.isFemale() ? this : void 0 : onlyDescendants ? void 0 : mother?.generateTree({
          generations: gens,
          level: level + 1,
          famIndex: famIndex * 2,
          index: index * 2 + (famIndex > 0 ? 0 : 1),
          ancestorSpouses,
          ancestorSiblings,
          maxLevel,
          minLevel,
          drawNonBiologicalAncestors
        }),
        motherSpouses: motherSpouses && (motherSpouses.values() ?? []),
        children: children?.map((descendant) => {
          const descendantTree = descendant.generateTree({
            generations: gens,
            level: level === 1 ? -1 : level - 1,
            famIndex: 0,
            index: childIndex++,
            showDescendants: true,
            offspringSpouses,
            maxLevel,
            minLevel
          });
          if (offspringSpouses) {
            const spousesTrees = descendant.getCoParents().orderBy(
              getMarriageAscAndChildBirth(descendant.id)
            ).map((spouse) => {
              return spouse.generateTree({
                generations: gens,
                level: level === 1 ? -1 : level - 1,
                famIndex: 0,
                index: childIndex++,
                showDescendants: true,
                offspringSpouses,
                maxLevel,
                minLevel
              });
            });
            return [descendantTree, ...spousesTrees];
          }
          return [descendantTree];
        }).flat()
      };
      if (!gens.tree[level]) {
        gens.tree[level] = [];
      }
      if (level < 0) {
        gens.tree[level].push(closers);
      } else {
        if (index >= half && gens.halves[level] === void 0) {
          gens.halves[level] = gens.tree[level].length;
        }
        gens.tree[level].push(closers);
        gens.lastItems[level] = closers;
      }
      if (father?.id) {
        gens.existed[father.id] = "yes" /* YES */;
      }
      if (mother?.id) {
        gens.existed[mother.id] = "yes" /* YES */;
      }
      closers.children?.forEach(({ id: id2 }) => {
        if (id2) {
          gens.existed[id2] = "yes" /* YES */;
        }
      });
      [
        ...closers.fatherSpouses ?? [],
        ...closers.motherSpouses ?? []
      ]?.forEach(({ id: id2 }) => {
        if (id2) {
          gens.existed[id2] = "spouse" /* SPOUSE */;
        }
      });
    });
    return this;
  }
  getPlaces(type = ["ALL" /* All */], maxLevel = 1) {
    if (!this._gedcom || !this.id) {
      return [];
    }
    const cacheKey = JSON.stringify(type);
    if (this._places?.[cacheKey]) {
      return this._places[cacheKey];
    }
    let marriages;
    if (type?.includes("MARR" /* Marriage */) || type?.includes("ALL" /* All */)) {
      const families = this.getFamilies("FAMS");
      if (families.length >= 1) {
        marriages = new List();
        families.forEach((fam) => {
          if (fam.MARR) {
            const newMarr = createCommon(
              this._gedcom,
              void 0,
              this
            );
            Object.assign(newMarr, fam.MARR);
            newMarr.id = fam.id;
            marriages.append(newMarr);
          }
        });
      }
    }
    let usedIndi;
    if (marriages) {
      usedIndi = createIndi(this._gedcom, this.id);
      Object.assign(usedIndi, this, {
        MARR: marriages
      });
    }
    const places = getPlaces(usedIndi || this, type, maxLevel);
    return places;
  }
  getTree(props) {
    const {
      descendants,
      offspringSpouses,
      ancestorSpouses,
      onlyDescendants,
      maxGen,
      minGen,
      drawNonBiologicalAncestors,
      ancestorSiblings
    } = props ?? {};
    const newGenerations = {
      existed: {},
      tree: {},
      halves: {},
      lastItems: {}
    };
    this.generateTree({
      generations: newGenerations,
      level: 1,
      famIndex: 0,
      index: 0,
      showDescendants: descendants,
      offspringSpouses,
      ancestorSpouses,
      onlyDescendants,
      maxLevel: maxGen,
      minLevel: minGen,
      drawNonBiologicalAncestors,
      ancestorSiblings
    });
    return Object.keys(newGenerations.tree).toSorted((a, b) => Number(a) - Number(b)).map((key) => {
      const gen = Number(key);
      return {
        gen,
        indis: newGenerations.tree[gen],
        half: newGenerations.halves[gen]
      };
    });
  }
  getGenealogy(onlyStraight = false, showDescendants = false) {
    const id = this.get("FAMC")?.index(0)?.toValue();
    if (!id) {
      return;
    }
    const own = this.isMale() ? {
      father: this,
      mother: !onlyStraight ? this.getCoParents().index(0) : void 0
    } : {
      mother: this,
      father: !onlyStraight ? this.getCoParents().index(0) : void 0
    };
    const newGenerations = {
      existed: {
        [this.id]: "yes" /* YES */
      },
      tree: {
        0: [
          {
            id,
            index: 0,
            ...own
          }
        ]
      },
      halves: { 0: 0 },
      lastItems: {}
    };
    this.generateTree({
      generations: newGenerations,
      level: 1,
      famIndex: 0,
      index: 0,
      showDescendants
    });
    const gens = Object.keys(newGenerations.tree).toSorted(
      (a, b) => Number(b) - Number(a)
    );
    const genealogyGenerations = {};
    const genealogyResult = {};
    gens.forEach((genIndex) => {
      const gen = Number(genIndex);
      const generation = newGenerations.tree[gen].flat();
      if (!genealogyGenerations[gen]) {
        genealogyGenerations[gen] = {
          left: [],
          main: { left: [], right: [] },
          right: []
        };
      } else {
        if (!genealogyGenerations[gen].left) {
          genealogyGenerations[gen].left = [];
        }
        if (!genealogyGenerations[gen].main) {
          genealogyGenerations[gen].main = { left: [], right: [] };
        }
        if (!genealogyGenerations[gen].right) {
          genealogyGenerations[gen].right = [];
        }
      }
      generation.forEach((pack, index) => {
        const members = { left: [], right: [] };
        ["father", "mother"].forEach((type, mIndex) => {
          const key = type;
          const typedPack = pack?.[key];
          const validIndex = index * 2 + mIndex;
          const side = validIndex >= generation.length ? "right" : "left";
          if (typedPack) {
            members[side].push(typedPack);
          }
        });
        if (!onlyStraight) {
          ["father", "mother"].forEach((type, mIndex) => {
            const key = type;
            const typedPack = pack?.[key];
            const validIndex = index * 2 + mIndex;
            const side = validIndex >= generation.length ? "right" : "left";
            const spouse = key === "father" ? pack?.mother : pack?.father;
            const spouses = (typedPack?.getCoParents().values() ?? []).filter((item) => {
              if (item?.id === spouse?.id) {
                return false;
              }
              if (item.id && !newGenerations.existed[item.id]) {
                newGenerations.existed[item.id] = "yes" /* YES */;
                return true;
              }
              return false;
            });
            const rawSiblings = typedPack?.getSiblings().orderBy(BIRTH_ASC).values() ?? [];
            const siblings = rawSiblings.reduce((acc, item, index2) => {
              if (item.id && !newGenerations.existed[item.id]) {
                newGenerations.existed[item.id] = "yes" /* YES */;
                const siblingSpouses = (item?.getCoParents().values() ?? []).filter((siblingSpouse) => {
                  if (siblingSpouse.id && !newGenerations.existed[siblingSpouse.id]) {
                    newGenerations.existed[siblingSpouse.id] = "yes" /* YES */;
                    return true;
                  }
                  return false;
                });
                if (siblingSpouses.length) {
                  const newValue = item?.isMale() ? [item, ...siblingSpouses] : [...siblingSpouses, item];
                  return [
                    ...acc,
                    ...index2 > 0 && newValue.length > 1 && acc[acc.length - 1] !== void 0 ? [void 0] : [],
                    ...newValue,
                    ...index2 < rawSiblings.length - 1 && newValue.length > 1 ? [void 0] : []
                  ];
                }
                return [...acc, item];
              }
              return acc;
            }, []);
            const spouseMethod = key === "father" ? "unshift" : "push";
            const siblingMethod = side === "left" ? "unshift" : "push";
            members[side][spouseMethod](...spouses);
            if (siblings.length) {
              if (key === "father" && pack?.mother || key === "mother" && pack?.father) {
                members[side][siblingMethod](void 0);
              }
              members[side][siblingMethod](...siblings);
            }
          });
        }
        genealogyGenerations[gen].main.left.push(
          members.left.length ? members.left : void 0
        );
        genealogyGenerations[gen].main.right.push(
          members.right.length ? members.right : void 0
        );
      });
    });
    gens.forEach((genIndex) => {
      const gen = Number(genIndex);
      const prevGen = gen - 1;
      if (!genealogyGenerations[prevGen]) {
        genealogyGenerations[prevGen] = {
          left: [],
          main: { left: [], right: [] },
          right: []
        };
      }
      ["left", "right"].forEach((s) => {
        const side = s;
        if (!onlyStraight) {
          genealogyGenerations[gen].main[side].forEach((pack) => {
            if (!pack) {
              return;
            }
            const toRight = side === "right";
            const orderedPack = pack;
            orderedPack.forEach((indi) => {
              if (!indi) {
                return;
              }
              const rawChildren = indi?.getChildren().orderBy(BIRTH_ASC).values() ?? [];
              const children = rawChildren.reduce((acc, item, index) => {
                if (item.id && !newGenerations.existed[item.id]) {
                  newGenerations.existed[item.id] = "yes" /* YES */;
                  const childSpouses = (item?.getCoParents().values() ?? []).filter((childSpouse) => {
                    if (childSpouse.id && !newGenerations.existed[childSpouse.id]) {
                      newGenerations.existed[childSpouse.id] = "yes" /* YES */;
                      return true;
                    }
                    return false;
                  });
                  if (childSpouses.length) {
                    const newValue = item?.isMale() ? [item, ...childSpouses] : [...childSpouses, item];
                    return [
                      ...acc,
                      ...index > 0 && newValue.length > 1 && acc[acc.length - 1] !== void 0 ? [void 0] : [],
                      ...newValue,
                      ...index < rawChildren.length - 1 && newValue.length > 1 ? [void 0] : []
                    ];
                  }
                  return [...acc, item];
                }
                return acc;
              }, []);
              const childMethod = "push";
              const members = [];
              if (children.length) {
                members[childMethod](children);
              }
              if (toRight) {
                genealogyGenerations[prevGen].right.push(
                  ...members
                );
              } else {
                genealogyGenerations[prevGen].left.push(
                  ...members
                );
              }
            });
          });
        }
        if (genealogyGenerations[prevGen]) {
          genealogyGenerations[prevGen] = {
            left: [],
            main: {
              left: [
                ...genealogyGenerations[prevGen]?.left ?? [],
                ...genealogyGenerations[prevGen]?.main.left ?? []
              ],
              right: [
                ...genealogyGenerations[prevGen]?.main.right ?? [],
                ...genealogyGenerations[prevGen]?.right ?? []
              ]
            },
            right: []
          };
        }
        genealogyResult[gen] = {
          left: [
            ...genealogyGenerations[gen]?.left ?? [],
            ...genealogyGenerations[gen]?.main.left ?? []
          ],
          right: [
            ...genealogyGenerations[gen]?.main.right ?? [],
            ...genealogyGenerations[gen]?.right ?? []
          ]
        };
      });
    });
    const result = [];
    gens.forEach((genIndex) => {
      result.unshift({
        gen: Number(genIndex),
        ...genealogyResult[Number(genIndex)]
      });
    });
    return result;
  }
  ancestryLink() {
    const www = this._gedcom?.HEAD?.SOUR?.CORP?.WWW?.value;
    const tree = this.getAncestryTreeId();
    if (this.id) {
      return `https://${www}/family-tree/person/tree/${tree}/person/${this.id.replace(
        /@|I/g,
        ""
      )}/facts`;
    }
  }
  async ancestryMedia(namespace) {
    const list = {};
    const objeList = this.get("OBJE")?.toList().copy();
    const www = this._gedcom?.HEAD?.SOUR?.CORP?.WWW?.value;
    const tree = this.getAncestryTreeId();
    if (objeList) {
      await Promise.all(
        objeList.map(async (objeRef) => {
          const key = objeRef?.id;
          const obje = objeRef?.standardizeMedia(
            namespace,
            true,
            (ns, iId) => {
              return ns && iId ? `https://mediasvc.ancestry.com/v2/image/namespaces/${ns}/media/${iId}?client=trees-mediaservice&imageQuality=hq` : void 0;
            }
          );
          const isPrimary = obje?.get("_PRIM")?.toValue() === "Y";
          const media = obje?.RIN?.toValue() || obje?.get("_OID")?.toValue();
          const clone = obje?.get("_CLON._OID")?.toValue();
          const mser = obje?.get("_MSER._LKID")?.toValue();
          let url = obje?.get("FILE")?.toValue();
          const title = obje?.get("TITL")?.toValue() ?? "";
          const type = obje?.get("FORM")?.toValue() ?? "raw";
          const imgId = clone || mser;
          if (!www || !tree || !this.id) {
            return;
          }
          if (!namespace && !url) {
            try {
              if (media) {
                const mediaDetailsResponse = await fetch(
                  `https://www.ancestry.com/api/media/viewer/v2/trees/${tree}/media?id=${media}`
                );
                const mediaDetails = await mediaDetailsResponse.json();
                if (mediaDetails.url) {
                  url = `${mediaDetails.url}&imageQuality=hq`;
                }
              }
            } catch (_e) {
            }
            url = url || `https://${www}/mediaui-viewer/tree/${tree}/media/${media}`;
          }
          if (url && imgId) {
            const id = `${tree}-${this.id}-${imgId}`;
            list[id] = {
              key,
              isPrimary,
              id,
              tree,
              imgId,
              person: this.id,
              title,
              url,
              contentType: type,
              downloadName: `${this.id.replaceAll("@", "")}_${this.toNaturalName()?.replaceAll(" ", "-") || ""}_${(title || key.replaceAll("@", "").toString()).replaceAll(" ", "-")}`
            };
          }
        })
      );
    }
    return list;
  }
  myheritageLink(poolId = 0) {
    const www = this._gedcom?.HEAD?.SOUR?.CORP?.value?.toLowerCase().replace(/^www\./gi, "");
    const site = this.getMyHeritageTreeId();
    const file = this._gedcom?.HEAD?.get("FILE")?.toValue()?.match(/Exported by MyHeritage.com from .+ in (?<site>.+) on .+$/)?.groups?.site;
    const normalizedFile = file?.normalize("NFD").replace(/[\u0300-\u036f]/g, "").toLowerCase().replace(/[^a-z0-9]+/g, "-");
    if (normalizedFile && this.id) {
      const id = Number(this.id.replace(/@|I/g, "")) + poolId;
      return `https://www.${www}/site-family-tree-${site}/${normalizedFile}#!profile-${id}-info`;
    }
  }
  myheritageMedia() {
    const list = {};
    const tree = this.getMyHeritageTreeId();
    if (!tree) {
      return;
    }
    const objeList = this.get("OBJE")?.toList().copy();
    const birthObj = this.get("BIRT.OBJE")?.toList().copy();
    const deathObj = this.get("DEAT.OBJE")?.toList().copy();
    objeList?.merge(birthObj).merge(deathObj);
    (this.get("FAMS")?.toValueList().values() ?? []).concat(this.get("FAMC")?.toValueList().values() ?? []).forEach((fam) => {
      objeList?.merge(fam?.get("MARR.OBJE"));
    });
    objeList?.forEach((o, index) => {
      if (!o) {
        return;
      }
      const obje = o;
      const key = `@O${index}@`;
      obje.standardizeMedia();
      const isPrimary = obje?.get("_PRIM")?.toValue() === "Y";
      const url = obje?.get("FILE")?.toValue();
      const title = obje?.get("NOTE")?.toValue() ?? "";
      const type = obje?.get("FORM")?.toValue() ?? "raw";
      const imgId = obje?.get("_PHOTO_RIN")?.toValue();
      if (url && imgId) {
        const id = `${tree}-${this.id}-${imgId}`;
        list[id] = {
          isPrimary,
          key,
          id,
          tree,
          imgId,
          person: this.id,
          title,
          url,
          contentType: type,
          downloadName: `${this.id.replaceAll("@", "")}_${this.toNaturalName()?.replaceAll(" ", "-") || ""}_${(title || key.replaceAll("@", "").toString()).replaceAll(" ", "-")}`
        };
      }
    });
    return list;
  }
  familySearchLink() {
    const wwwTags = this.get("WWW")?.toList();
    const wwwTag = wwwTags?.find((wwwTag2) => {
      const fsSourceMarker = wwwTag2.get("_IS_FS")?.toValue();
      return fsSourceMarker === "Y";
    })?.toValue();
    if (wwwTag) {
      return wwwTag;
    }
    return this.get("_FS_LINK")?.toValue();
  }
  hasFamilySearchMatches() {
    const matchTags = this.get("MATCH")?.toList();
    if (matchTags && matchTags.length > 0) {
      return true;
    }
    return this.get("_FS_MATCH") !== void 0;
  }
  getFamilySearchMatches() {
    let matchTags = this.get("MATCH")?.toList();
    if (!matchTags?.length) {
      matchTags = this.get("_FS_MATCH")?.toList();
    }
    if (!matchTags?.length) {
      return [];
    }
    return matchTags.map((matchTag) => {
      const id = matchTag.toValue();
      const title = matchTag.get("TITL")?.toValue();
      const type = matchTag.get("TYPE")?.toValue();
      const ref = matchTag.get("REF")?.toValue();
      const scoreStr = matchTag.get("SCORE")?.toValue();
      const text = matchTag.get("TEXT")?.toValue();
      const www = matchTag.get("WWW")?.toValue();
      const noteTags = matchTag.get("NOTE")?.toList() || [];
      const notes = noteTags.map((note) => note.toValue()).filter(Boolean);
      return {
        id,
        title,
        type,
        ref,
        score: scoreStr ? parseFloat(scoreStr) : void 0,
        text,
        www,
        notes
      };
    });
  }
  hasFamilySearchSources() {
    const sourTags = this.get("SOUR")?.toList();
    const sourTag = sourTags?.find((sourTag2) => {
      const fsSourceMarker = sourTag2.get("_IS_FS")?.toValue();
      return fsSourceMarker === "Y";
    });
    if (sourTag) {
      return true;
    }
    return this.get("_FS_SOUR") !== void 0;
  }
  getFamilySearchSources() {
    let sourceTags = this.get("SOUR")?.toList()?.filter((sourTag) => {
      const fsSourceMarker = sourTag.get("_IS_FS")?.toValue();
      return fsSourceMarker === "Y";
    });
    if (!sourceTags || !sourceTags.length) {
      sourceTags = this.get("_FS_SOUR")?.toList();
    }
    if (!sourceTags || !sourceTags.length) {
      return [];
    }
    return sourceTags.map((sourceTag) => {
      const id = sourceTag.toValue();
      const title = sourceTag.get("TITL")?.toValue();
      const text = sourceTag.get("TEXT")?.toValue();
      const www = sourceTag.get("WWW")?.toValue();
      const noteTags = sourceTag.get("NOTE")?.toList() || [];
      const notes = noteTags.map((note) => note.toValue()).filter(Boolean);
      return {
        id,
        title,
        text,
        www,
        notes
      };
    });
  }
  geniMedia() {
    const list = {};
    const objeList = this.get("OBJE")?.toList().copy();
    const sourList = this.get("SOUR")?.toList().copy();
    sourList?.forEach((sour) => {
      const sourObje = sour?.get("OBJE")?.toList();
      objeList?.merge(sourObje);
    });
    const rfn = this.get("RFN")?.toValue();
    const geniId = rfn?.replace(/^geni:/, "") || "unknown";
    objeList?.forEach((obje, index) => {
      if (!obje) {
        return;
      }
      const key = `@O${index}@`;
      const isPrimary = obje?.get("_PRIM")?.toValue() === "Y";
      const url = obje?.get("FILE")?.toValue();
      const title = obje?.get("TITL")?.toValue() ?? "";
      const type = obje?.get("FORM")?.toValue() ?? "raw";
      if (url) {
        const urlMatch = url.match(/\/([^/]+)\?hash=/);
        const imgId = urlMatch?.[1] || `img-${index}-${Date.now().toString(36)}`;
        const id = `geni-${geniId}-${imgId}`;
        list[id] = {
          isPrimary,
          key,
          id,
          tree: geniId,
          imgId,
          person: this.id,
          title,
          url,
          contentType: type,
          downloadName: `${this.id.replaceAll("@", "")}_${this.toNaturalName()?.replaceAll(" ", "-") || ""}_${(title || key.replaceAll("@", "").toString()).replaceAll(" ", "-")}`
        };
      }
    });
    return list;
  }
  universalMedia() {
    const list = {};
    if (!this.id) {
      return list;
    }
    const objeList = this.get("OBJE")?.toList().copy();
    if (!objeList || objeList.length === 0) {
      return list;
    }
    const rfn = this.get("RFN")?.toValue();
    const treeId = this.getUniversalTreeId() || rfn || "universal";
    objeList.forEach((obje, index) => {
      if (!obje) {
        return;
      }
      const key = `@O${index}@`;
      obje.standardizeMedia();
      const isPrimary = obje?.get("_PRIM")?.toValue() === "Y";
      const url = obje?.get("FILE")?.toValue();
      const title = obje?.get("TITL")?.toValue() ?? "";
      const type = obje?.get("FORM")?.toValue() ?? "raw";
      if (url) {
        const imgId = `media-${index}-${url.split("/").pop()?.split("?")[0]?.substring(0, 20) || Date.now().toString(36)}`;
        const id = `${treeId}-${this.id}-${imgId}`;
        list[id] = {
          isPrimary,
          key,
          id,
          tree: treeId,
          imgId,
          person: this.id,
          title,
          url,
          contentType: type,
          downloadName: `${this.id.replaceAll("@", "")}_${this.toNaturalName()?.replaceAll(" ", "-") || ""}_${(title || key.replaceAll("@", "").toString()).replaceAll(" ", "-")}`
        };
      }
    });
    return list;
  }
  async multimedia(namespace) {
    if (this?.isAncestry()) {
      return await this.ancestryMedia(namespace);
    }
    if (this?.isMyHeritage()) {
      return this.myheritageMedia();
    }
    if (this?.isGeni()) {
      return this.geniMedia();
    }
    return this.universalMedia();
  }
  async getProfilePicture(namespace, onlyPrimary = true) {
    if (!this.id) {
      return void 0;
    }
    const cacheKey = this.id;
    const cached = profilePictureCache(
      this._gedcom,
      cacheKey
    );
    if (cached !== void 0) {
      return cached;
    }
    const mediaList = await this.multimedia(namespace);
    if (!mediaList) {
      profilePictureCache(
        this._gedcom,
        cacheKey,
        void 0
      );
      return void 0;
    }
    const mediaArray = Object.values(mediaList);
    const primaryMedia = mediaArray.find(
      (media) => media.isPrimary && isImageFormat(media.contentType || getFileExtension(media.url))
    );
    if (primaryMedia) {
      const result = {
        file: primaryMedia.url,
        form: primaryMedia.contentType,
        title: primaryMedia.title,
        isPrimary: true
      };
      profilePictureCache(this._gedcom, cacheKey, result);
      return result;
    }
    if (onlyPrimary) {
      profilePictureCache(
        this._gedcom,
        cacheKey,
        void 0
      );
      return void 0;
    }
    const secondaryMedia = mediaArray.find(
      (media) => isImageFormat(media.contentType || getFileExtension(media.url))
    );
    if (secondaryMedia) {
      const result = {
        file: secondaryMedia.url,
        form: secondaryMedia.contentType,
        title: secondaryMedia.title,
        isPrimary: false
      };
      profilePictureCache(this._gedcom, cacheKey, result);
      return result;
    }
    profilePictureCache(
      this._gedcom,
      cacheKey,
      void 0
    );
    return void 0;
  }
  link(poolId) {
    if (this?.isAncestry()) {
      return this.ancestryLink();
    }
    if (this?.isMyHeritage()) {
      return this.myheritageLink(poolId);
    }
    if (this?.isFamilySearch()) {
      return this.familySearchLink();
    }
    return void 0;
  }
  toFamilies(list) {
    const families = new Families();
    list?.forEach((fam, famId) => {
      const family = fam?.ref;
      if (family) {
        families.item(famId, family);
      }
    });
    return families;
  }
  getFamilies(type) {
    return this.toFamilies(this.get(type)?.toValueList());
  }
  getFamiliesBiologicalFirst(type) {
    const families = this.toFamilies(this.get(type)?.toValueList());
    return families.orderBy((famB, _famBKey, famA) => {
      const husbA = famA.getHusband().index(0);
      const wifeA = famA.getWife().index(0);
      const husbAType = husbA && this.getParentType(husbA);
      const wifeAType = wifeA && this.getParentType(wifeA);
      const husbB = famB.getHusband().index(0);
      const wifeB = famB.getWife().index(0);
      const husbBType = husbB && this.getParentType(husbB);
      const wifeBType = wifeB && this.getParentType(wifeB);
      if (husbAType === "biological" /* BIOLOGICAL */ && wifeAType === "biological" /* BIOLOGICAL */) {
        if (husbBType === "biological" /* BIOLOGICAL */ && wifeBType === "biological" /* BIOLOGICAL */) {
          return 0;
        }
        return 1;
      } else if (husbAType === "biological" /* BIOLOGICAL */) {
        if (husbBType === "biological" /* BIOLOGICAL */ && wifeBType === "biological" /* BIOLOGICAL */) {
          return -1;
        } else if (husbBType === "biological" /* BIOLOGICAL */) {
          return 0;
        }
        return 1;
      } else if (wifeAType === "biological" /* BIOLOGICAL */) {
        if (husbBType === "biological" /* BIOLOGICAL */ && wifeBType === "biological" /* BIOLOGICAL */) {
          return -1;
        } else if (husbBType === "biological" /* BIOLOGICAL */) {
          return -1;
        }
        return 0;
      }
      if (husbBType === "biological" /* BIOLOGICAL */ || wifeBType === "biological" /* BIOLOGICAL */) {
        return -1;
      }
      return 0;
    });
  }
  getBirthDate(showDays = false, shortNote = true, showNote = true) {
    const dates = dateFormatter(
      this,
      false,
      showDays,
      false,
      shortNote,
      showNote
    );
    return dates.birth || void 0;
  }
  getDeathDate(showDays = false, shortNote = true, showNote = true) {
    const dates = dateFormatter(
      this,
      false,
      showDays,
      false,
      shortNote,
      showNote
    );
    return dates.death || void 0;
  }
  getBirthPlace() {
    const birthEvent = this.get("BIRT")?.index(0);
    return birthEvent?.PLAC?.value;
  }
  getDeathPlace() {
    const deathEvent = this.get("DEAT")?.index(0);
    return deathEvent?.PLAC?.value;
  }
  /**
   * Get all marriage places for this individual
   * @returns Array of marriage place strings (may include undefined for marriages without place)
   */
  getMarriagePlaces() {
    const marriagePlaces = [];
    const families = this.getFamilies("FAMS");
    families.forEach((family) => {
      const marrEvents = (family?.MARR?.toList().values() ?? []).filter(
        Boolean
      );
      if (marrEvents.length === 0) {
        marriagePlaces.push(void 0);
      } else {
        marrEvents.forEach((marrEvent) => {
          const marriageEventDetail = marrEvent;
          const marriagePlace = marriageEventDetail?.PLAC?.value;
          marriagePlaces.push(marriagePlace);
        });
      }
    });
    return marriagePlaces;
  }
  /**
   * Get all marriage dates for this individual
   * @returns Array of marriage date strings (may include undefined for marriages without date)
   */
  getMarriageDates() {
    const marriageDates = [];
    const families = this.getFamilies("FAMS");
    families.forEach((family) => {
      const marrEvents = (family?.MARR?.toList().values() ?? []).filter(
        Boolean
      );
      if (marrEvents.length === 0) {
        marriageDates.push(void 0);
      } else {
        marrEvents.forEach((marrEvent) => {
          const marriageEventDetail = marrEvent;
          const marriageDate = marriageEventDetail?.DATE?.value;
          marriageDates.push(marriageDate);
        });
      }
    });
    return marriageDates;
  }
  isDead() {
    return this.get("DEAT.DATE")?.toValue() !== void 0 || this.get("DEAT.PLAC")?.toValue() !== void 0;
  }
  isEmpty() {
    if (this._isEmpty !== void 0) {
      return this._isEmpty;
    }
    const keysLength = Object.keys(this).filter((key) => {
      return !key.startsWith("_") && !["gedcom", "isListable", "refType"].includes(key);
    }).length;
    this._isEmpty = !keysLength;
    return this._isEmpty;
  }
  isUnknownAncestor() {
    if (this._isUnknownAncestor !== void 0) {
      return this._isUnknownAncestor;
    }
    this._isUnknownAncestor = !!this.get("_MTTAG")?.toList()?.values().find(
      (tag) => tag?.get("NAME")?.toValue() === "Unknown Ancestor" /* UnknownAncestor */
    );
    return this._isUnknownAncestor;
  }
  isIgnoredMember() {
    if (this._isIgnoredPerson !== void 0) {
      return this._isIgnoredPerson;
    }
    this._isIgnoredPerson = !!this.get("_MTTAG")?.toList()?.values().find(
      (tag) => tag?.get("NAME")?.toValue() === "Ignored member" /* IgnoredMember */
    );
    return this._isIgnoredPerson;
  }
  isUnattachedMember() {
    if (this._isUnattachedMember !== void 0) {
      return this._isUnattachedMember;
    }
    this._isUnattachedMember = !!this.get("_MTTAG")?.toList()?.values().find(
      (tag) => tag?.get("NAME")?.toValue() === "Unattached member" /* UnattachedMember */
    );
    return this._isUnattachedMember;
  }
  isOrphanFamilyMember() {
    if (this._isOrphanFamilyMember !== void 0) {
      return this._isOrphanFamilyMember;
    }
    const famcFamilies = this.getFamilies("FAMC");
    const famsFamilies = this.getFamilies("FAMS");
    this._isOrphanFamilyMember = false;
    famcFamilies.forEach((fam) => {
      if (fam._IS_ORPHAN_FAMILY?.value === "Y") {
        this._isOrphanFamilyMember = true;
      }
    });
    if (!this._isOrphanFamilyMember) {
      famsFamilies.forEach((fam) => {
        if (fam._IS_ORPHAN_FAMILY?.value === "Y") {
          this._isOrphanFamilyMember = true;
        }
      });
    }
    return this._isOrphanFamilyMember;
  }
  isUnknownGivenname() {
    if (this._isUnknownGivenname !== void 0) {
      return this._isUnknownGivenname;
    }
    this._isUnknownGivenname = !!this.get("_MTTAG")?.toList()?.values().find(
      (tag) => tag?.get("NAME")?.toValue() === "Unknown givenname" /* UnknownGivenname */
    );
    return this._isUnknownGivenname;
  }
  isUnknownSurname() {
    if (this._isUnknownSurname !== void 0) {
      return this._isUnknownSurname;
    }
    this._isUnknownSurname = !!this.get("_MTTAG")?.toList()?.values().find(
      (tag) => tag?.get("NAME")?.toValue() === "Unknown surname" /* UnknownSurname */
    );
    return this._isUnknownSurname;
  }
  isNonRelevantMember() {
    return this.isUnknownAncestor() || this.isUnattachedMember();
  }
  isMale() {
    return this.get("SEX")?.toValue() === "M";
  }
  isFemale() {
    return this.get("SEX")?.toValue() === "F";
  }
  isUnknownSex() {
    return !this.isMale() && !this.isFemale();
  }
  getParentType(id) {
    let indi;
    if (typeof id === "string" || typeof id === "number") {
      indi = this._gedcom?.indi(id);
    } else {
      indi = id;
    }
    if (!indi?.id || !this.id) {
      return;
    }
    const parents = this.getParents();
    let parent = parents.item(indi.id) ? indi : void 0;
    let child = this._gedcom?.indi(this.id);
    if (!parent?.id) {
      const children = this.getChildren();
      if (children.item(indi.id)) {
        parent = this._gedcom?.indi(this.id);
        child = indi;
      }
    }
    const familiesOfChildren = child?.get("FAMC")?.toValueList();
    if (!parent?.id || !child?.id || !familiesOfChildren) {
      return;
    }
    const families = this.toFamilies(familiesOfChildren);
    let childType = "";
    families.forEach((family) => {
      if (childType) {
        return;
      }
      const fatherId = family.get("HUSB")?.toValue();
      const motherId = family.get("WIFE")?.toValue();
      const usedRel = parent?.id === fatherId ? "_FREL" : parent?.id === motherId ? "_MREL" : void 0;
      if (!usedRel || !child?.id) {
        return;
      }
      const famChild = this.id && family.CHIL?.toValueList().item(child.id);
      if (!famChild) {
        const childIndi = this._gedcom?.indi(child.id);
        const childState = childIndi?.get("FAMC")?.toList()?.filter((item) => item.value === family.id).index(0);
        childType = childState?.get("PEDI")?.toValue() || "biological" /* BIOLOGICAL */;
      } else {
        childType = famChild.get(usedRel)?.toValue() || "biological" /* BIOLOGICAL */;
      }
    });
    return (childType || "biological" /* BIOLOGICAL */).toLowerCase();
  }
  hasFacts() {
    const dates = dateFormatter(this, true);
    if (dates.inArray.length) {
      return true;
    }
    const facts = this.getFacts(1);
    return !!facts.length;
  }
  getLinks() {
    return this.get("_WLNK")?.toList();
  }
  getAkas(limit) {
    return this.getFacts(limit, "AKA");
  }
  getFacts(limit, filter) {
    const filters = (Array.isArray(filter) ? filter : [filter]).filter(
      Boolean
    );
    const facts = new List();
    let id = 0;
    ALLOWED_FACTS.forEach((fact) => {
      const isCustom = ["EVEN", "FACT"].includes(fact);
      if (filters.length && !filters.includes(fact) && !isCustom) {
        return;
      }
      const factCommon = this.get(fact);
      const factCommons = factCommon?.toList();
      factCommons?.forEach((common, _, index) => {
        if (limit !== void 0 && index >= limit) {
          return;
        }
        if (!common) {
          return;
        }
        const newCommon = createCommon(
          this._gedcom,
          `${id}`,
          this
        );
        Object.assign(newCommon, common, { id });
        const type = isCustom ? newCommon.get("TYPE")?.toValue() : fact;
        if (!type || DISALLOWED_CUSTOM_FACTS.includes(type)) {
          return;
        }
        if (filters.length && !filters.includes(type) && isCustom) {
          return;
        }
        const label = createCommon(this._gedcom, void 0, this);
        const customRenderer = CustomFactRenderers[type];
        if (customRenderer) {
          customRenderer(label, newCommon, this);
        } else {
          label.value = type;
          newCommon.set("_LABEL", label);
        }
        id++;
        facts.append(newCommon);
      });
    });
    return facts.orderBy(DATE_ASC);
  }
  commonAncestor(person, options) {
    const path = this.path(person, options);
    let wentUp = false;
    let lastItem;
    let ancestor;
    path?.forEach((item) => {
      if (lastItem) {
        if (item.level > lastItem.level) {
          wentUp = true;
        }
        if (item.level < lastItem.level) {
          wentUp = true;
          if (wentUp && !ancestor) {
            ancestor = lastItem.indi;
          }
        }
      }
      lastItem = item;
    });
    return ancestor;
  }
  path(person, options) {
    const { breakAfterSpouse = true, breakAfterNonBiological } = options ?? {};
    const usedIndi = typeof person === "string" ? this._gedcom?.indi(person) : person;
    if (!this.id || !usedIndi?.id) {
      return;
    }
    const cacheKey = `${this.id}|${usedIndi.id}`;
    const cache = pathCache(this._gedcom, cacheKey);
    if (cache) {
      return cache;
    }
    const visited = new Individuals().append(this);
    const mainItem = {
      indi: this,
      level: 0,
      levelUp: 0,
      levelDown: 0,
      degree: 0,
      kinship: "self"
    };
    const path = [mainItem];
    if (this.id === usedIndi.id) {
      return path;
    }
    const queue = [
      {
        ...mainItem,
        path
      }
    ];
    let helper = 0;
    while (queue.length > 0) {
      if (helper++ > 1e6) {
        break;
      }
      const {
        indi,
        path: path2,
        kinship,
        relation,
        level,
        levelUp,
        levelDown,
        degree,
        breakOnNext,
        breakAfterNext,
        inLaw
      } = queue.shift();
      if (usedIndi.id === indi.id) {
        if (breakOnNext) {
          return void 0;
        }
        pathCache(this._gedcom, cacheKey, path2);
        return path2;
      }
      visited.append(indi);
      const additional = {};
      if (breakOnNext || breakAfterNext) {
        additional.breakOnNext = breakOnNext || breakAfterNext;
      }
      if (inLaw) {
        additional.inLaw = inLaw;
      }
      if (kinship === "spouse" && breakAfterSpouse) {
        if (path2.length <= 2) {
          additional.inLaw = true;
        } else {
          additional.breakOnNext = true;
        }
      }
      if (kinship !== "child" || !breakAfterSpouse) {
        indi.getBiologicalFathers().copy().merge(indi.getBiologicalMothers()).merge(indi.getFathers().copy().merge(indi.getMothers())).forEach((relative) => {
          if (!visited.has(relative)) {
            const currentRelation = indi.getParentType(relative);
            if (breakAfterNonBiological && currentRelation !== "biological" /* BIOLOGICAL */) {
              additional.breakAfterNext = true;
            }
            const newItem = {
              indi: relative,
              kinship: "parent",
              relation: currentRelation && currentRelation !== "biological" /* BIOLOGICAL */ ? currentRelation : relation,
              level: level + 1,
              levelUp: levelUp + 1,
              levelDown,
              degree,
              ...additional
            };
            queue.push({
              ...newItem,
              path: [...path2, newItem]
            });
          }
        });
      }
      indi.getChildren().forEach((relative) => {
        if (!visited.has(relative)) {
          const currentRelation = relative.getParentType(indi);
          if (breakAfterNonBiological && currentRelation !== "biological" /* BIOLOGICAL */) {
            additional.breakAfterNext = true;
          }
          const newItem = {
            indi: relative,
            kinship: "child",
            relation: currentRelation && currentRelation !== "biological" /* BIOLOGICAL */ ? currentRelation : relation,
            level: level - 1,
            levelUp,
            levelDown: levelDown + 1,
            degree: levelUp ? level > 0 ? levelUp - level + 1 : levelDown - Math.abs(level) : 0,
            ...additional
          };
          queue.push({
            ...newItem,
            path: [...path2, newItem]
          });
        }
      });
      indi.getCoParents().forEach((relative) => {
        if (!visited.has(relative)) {
          const currentAddition = {};
          if (relation && relation !== "biological" /* BIOLOGICAL */) {
            currentAddition.relation = relation;
          }
          if (inLaw) {
            currentAddition.breakOnNext = true;
          }
          const newItem = {
            indi: relative,
            kinship: "spouse",
            level,
            levelUp,
            levelDown,
            degree,
            ...additional,
            ...currentAddition
          };
          queue.push({
            ...newItem,
            path: [...path2, newItem]
          });
        }
      });
    }
    return void 0;
  }
  kinship(other, showMainPerson, lang = "en", entirePath, displayName = "givenname") {
    const KinshipTranslatorClass2 = getKinshipTranslatorClass();
    const translator = new KinshipTranslatorClass2(
      this,
      other,
      lang,
      entirePath,
      showMainPerson ? displayName : void 0
    );
    return translator.translate(!!showMainPerson);
  }
  isRelativeOf(type, indi) {
    const usedIndi = typeof indi === "string" ? this._gedcom?.indi(indi) : indi;
    let getter;
    if (type === "fullsibling") {
      getter = "getFullSiblings";
    }
    if (type === "halfsibling") {
      getter = "getHalfSiblings";
    }
    if (type === "sibling") {
      getter = "getSiblings";
    }
    if (type === "parent") {
      getter = "getParents";
    }
    if (type === "child") {
      getter = "getChildren";
    }
    if (type === "spouse") {
      getter = "getCoParents";
    }
    if (type === "siblingInLaw") {
      getter = "getSiblingsInLaw";
    }
    if (type === "parentInLaw") {
      getter = "getParentsInLaw";
    }
    if (type === "childInLaw") {
      getter = "getChildrenInLaw";
    }
    if (!usedIndi || !getter) {
      return false;
    }
    const relatives = usedIndi[getter]();
    return Boolean(
      this.id && relatives.item(this.id) ? relatives.index(0)?.id || true : false
    );
  }
  isSiblingOf(indi) {
    return this.isRelativeOf("sibling", indi);
  }
  isFullSiblingOf(indi) {
    return this.isRelativeOf("fullsibling", indi);
  }
  isHalfSiblingOf(indi) {
    return this.isRelativeOf("halfsibling", indi);
  }
  isSpouseOf(indi) {
    return this.isRelativeOf("spouse", indi);
  }
  isParentOf(indi) {
    return this.isRelativeOf("parent", indi);
  }
  isChildOf(indi) {
    return this.isRelativeOf("child", indi);
  }
  isSiblingInLawOf(indi) {
    return this.isRelativeOf("siblingInLaw", indi);
  }
  isParentInLawOf(indi) {
    return this.isRelativeOf("parentInLaw", indi);
  }
  isChildInLawOf(indi) {
    return this.isRelativeOf("childInLaw", indi);
  }
  getRelativesOnDegree(degree = 0) {
    this.id = this.id || `@I${Math.random()}@`;
    const cache = relativesOnDegreeCache(this._gedcom, this.id, degree);
    if (cache) {
      return cache;
    }
    let persons = this.getRelativesOnLevel(1).getRelativesOnDegree(-1).copy();
    const excludes = persons;
    if (!Math.abs(degree)) {
      return relativesOnDegreeCache(
        this._gedcom,
        this.id,
        degree,
        persons.except(this)
      );
    }
    for (let i = 1; i < Math.abs(degree) + 1; i++) {
      const validDegree = i + 1;
      excludes.merge(persons);
      persons = this.getRelativesOnLevel(validDegree).getRelativesOnDegree(-validDegree).copy().exclude(excludes);
    }
    return relativesOnDegreeCache(this._gedcom, this.id, degree, persons);
  }
  getRelativesOnLevel(level = 0, filter) {
    this.id = this.id || `@I${Math.random()}@`;
    const cache = relativesOnLevelCache(this._gedcom, this.id, level);
    if (cache) {
      return cache;
    }
    let persons = new Individuals();
    const config = {
      isAscendant: level < 0,
      key: level <= 0 ? "FAMS" : "FAMC"
    };
    let families = this.get(config.key)?.toValueList();
    if (!families) {
      return relativesOnLevelCache(this._gedcom, this.id, level, persons);
    }
    if (filter) {
      families = families.filter(filter);
    }
    if (config.isAscendant) {
      persons = this.toFamilies(families).getChildren();
    } else {
      persons = this.toFamilies(families).getParents();
    }
    if (level >= -1 && level <= 1) {
      return relativesOnLevelCache(
        this._gedcom,
        this.id,
        level,
        persons.except(this)
      );
    }
    for (let i = 1; i < Math.abs(level); i++) {
      if (config.isAscendant) {
        persons = persons.getChildren();
      } else {
        persons = persons.getParents();
      }
    }
    return relativesOnLevelCache(
      this._gedcom,
      this.id,
      level,
      persons.except(this)
    );
  }
  getAscendants(level = 0, filter) {
    if (!level) {
      return new Individuals();
    }
    return this.getRelativesOnLevel(level, filter);
  }
  getDescendants(level = 0, filter) {
    if (!level) {
      return new Individuals();
    }
    return this.getRelativesOnLevel(-level, filter);
  }
  getAllDescendantsRaw(individuals, containDescendantsInLaw = false) {
    this.id = this.id || `@I${Math.random()}@`;
    let ownGeneration = new Individuals();
    if (individuals) {
      ownGeneration = individuals;
      ownGeneration.merge(this.getCoParents());
    }
    ownGeneration.append(this);
    const relatives = new Individuals();
    relatives.merge(ownGeneration);
    const generations = {
      0: ownGeneration
    };
    let currentGen = 0;
    const maxGenCheck = 100;
    while (currentGen < maxGenCheck) {
      const descentants = generations[currentGen]?.getChildren().copy();
      if (!descentants?.length) {
        break;
      }
      if (containDescendantsInLaw) {
        descentants?.merge(descentants?.getCoParents());
      }
      currentGen++;
      generations[currentGen] = descentants;
      relativesOnLevelCache(
        this._gedcom,
        this.id,
        -currentGen,
        descentants
      );
      descentants && relatives.merge(descentants);
    }
    return { relatives, generations };
  }
  getAllDescendants(individuals, containDescendantsInLaw = false) {
    const raw = this.getAllDescendantsRaw(
      individuals,
      containDescendantsInLaw
    );
    return raw.relatives;
  }
  getAllAscendantsRaw(individuals) {
    this.id = this.id || `@I${Math.random()}@`;
    let ownGeneration = new Individuals();
    if (individuals) {
      ownGeneration = individuals;
    }
    ownGeneration.append(this);
    const relatives = new Individuals();
    relatives.merge(ownGeneration);
    const generations = {
      0: ownGeneration
    };
    let currentGen = 0;
    const maxGenCheck = 100;
    while (currentGen < maxGenCheck) {
      const parents = generations[currentGen]?.getParents().copy();
      if (!parents?.length) {
        break;
      }
      currentGen++;
      generations[currentGen] = parents;
      relativesOnLevelCache(this._gedcom, this.id, currentGen, parents);
      parents && relatives.merge(parents);
    }
    return { relatives, generations };
  }
  getAllAscendants(individuals) {
    const raw = this.getAllAscendantsRaw(individuals);
    return raw.relatives;
  }
  getHalfSiblings() {
    const siblings = this.getSiblings();
    const ownParents = this.getBiologicalParents();
    const halfSiblings = new Individuals();
    siblings.forEach((sibling) => {
      const siblingsParents = sibling.getBiologicalParents();
      const inter = ownParents.intersection(siblingsParents);
      if (inter.length < ownParents.length) {
        halfSiblings.append(sibling);
      }
    });
    return halfSiblings;
  }
  getFullSiblings() {
    const siblings = this.getSiblings();
    const ownParents = this.getBiologicalParents();
    const fullSiblings = new Individuals();
    siblings.forEach((sibling) => {
      const siblingsParents = sibling.getBiologicalParents();
      const inter = ownParents.intersection(siblingsParents);
      if (inter.length === ownParents.length) {
        fullSiblings.append(sibling);
      }
    });
    return fullSiblings;
  }
  getSiblings() {
    implemented();
    return this.getRelativesOnDegree(0);
  }
  getBrothers() {
    implemented();
    return this.getSiblings().filter(MALE);
  }
  getSisters() {
    implemented();
    return this.getSiblings().filter(MALE);
  }
  getChildren(filter) {
    implemented();
    return this.getDescendants(1, filter);
  }
  getChildrenFilteredByPedigree(filter) {
    const children = new Individuals();
    const familiesOfChildrens = this.get("FAMS")?.toValueList();
    if (!familiesOfChildrens || !this.id) {
      return children;
    }
    const families = this.toFamilies(familiesOfChildrens);
    families.forEach((family) => {
      const fatherId = family.get("HUSB")?.toValue();
      const motherId = family.get("WIFE")?.toValue();
      const usedRel = this.id === fatherId ? "_FREL" : this.id === motherId ? "_MREL" : void 0;
      if (!usedRel) {
        return;
      }
      const famChildren = this.id && family.CHIL?.toValueList();
      famChildren?.forEach((child) => {
        const childType = child.get(usedRel)?.toValue();
        if (!childType && filter.PEDI === "biological" /* BIOLOGICAL */ || childType?.toLowerCase() === filter.PEDI.toLowerCase()) {
          const childId = child?.toValue();
          const childIndi = childId && this._gedcom?.indi(childId);
          if (childIndi) {
            children.append(childIndi);
          }
        }
      });
      if (!famChildren?.length) {
        this.getChildren().forEach((child) => {
          if (!child) {
            return;
          }
          const childState = child?.get("FAMC")?.toList()?.filter((item) => item.value === family.id).index(0);
          const childType = childState?.get("PEDI")?.toValue() || "biological" /* BIOLOGICAL */;
          if (childType?.toLowerCase() === filter.PEDI.toLowerCase()) {
            children.append(child);
          }
        });
      }
    });
    return children;
  }
  getBiologicalChildren() {
    implemented();
    return this.getChildrenFilteredByPedigree(BIOLOGICAL);
  }
  getAdoptedChildren() {
    implemented();
    return this.getChildrenFilteredByPedigree(ADOPTED);
  }
  getBirthChildren() {
    implemented();
    return this.getChildrenFilteredByPedigree(BIRTH);
  }
  getFosterChildren() {
    implemented();
    return this.getChildrenFilteredByPedigree(FOSTER);
  }
  getSealingChildren() {
    implemented();
    return this.getChildrenFilteredByPedigree(SEALING);
  }
  getStepChildren() {
    implemented();
    return this.getChildrenFilteredByPedigree(STEP);
  }
  getSons() {
    implemented();
    return this.getChildren().filter(MALE);
  }
  getBiologicalSons() {
    implemented();
    return this.getChildrenFilteredByPedigree(BIOLOGICAL).filter(
      MALE
    );
  }
  getAdoptedSons() {
    implemented();
    return this.getAdoptedChildren().filter(MALE);
  }
  getBirthSons() {
    implemented();
    return this.getBirthChildren().filter(MALE);
  }
  getFosterSons() {
    implemented();
    return this.getFosterChildren().filter(MALE);
  }
  getSealingSons() {
    implemented();
    return this.getSealingChildren().filter(MALE);
  }
  getStepSons() {
    implemented();
    return this.getStepChildren().filter(MALE);
  }
  getDaughters() {
    implemented();
    return this.getChildren().filter(FEMALE);
  }
  getBiologicalDaugthers() {
    implemented();
    return this.getChildrenFilteredByPedigree(BIOLOGICAL).filter(
      FEMALE
    );
  }
  getAdoptedDaughters() {
    implemented();
    return this.getAdoptedChildren().filter(FEMALE);
  }
  getBirthDaughters() {
    implemented();
    return this.getBirthChildren().filter(FEMALE);
  }
  getFosterDaughters() {
    implemented();
    return this.getFosterChildren().filter(FEMALE);
  }
  getSealingDaughters() {
    implemented();
    return this.getSealingChildren().filter(FEMALE);
  }
  getStepDaughters() {
    implemented();
    return this.getStepChildren().filter(FEMALE);
  }
  getParents(filter) {
    implemented();
    return this.getAscendants(1, filter);
  }
  getParentsFilteredByPedigree(filter) {
    const parents = new Individuals();
    const familiesOfParents = this.get("FAMC")?.toValueList();
    if (!familiesOfParents || !this.id) {
      return parents;
    }
    const families = this.toFamilies(familiesOfParents);
    families.forEach((family) => {
      const child = this.id && family.CHIL?.toValueList()?.item(this.id);
      if (!child) {
        const childState = this?.get("FAMC")?.toList()?.filter((item) => item.value === family.id).index(0);
        const parentType = childState?.get("PEDI")?.toValue() || "biological" /* BIOLOGICAL */;
        if (parentType?.toLowerCase() === filter.PEDI.toLowerCase()) {
          parents.merge(family.getParents());
        }
        return;
      }
      const fatherType = child.get("_FREL")?.toValue();
      const motherType = child.get("_MREL")?.toValue();
      if (!fatherType && filter.PEDI === "biological" /* BIOLOGICAL */ || fatherType?.toLowerCase() === filter.PEDI.toLowerCase()) {
        const fatherId = family.get("HUSB")?.toValue();
        const father = fatherId && this._gedcom?.indi(fatherId);
        if (father) {
          parents.append(father);
        }
      }
      if (!motherType && filter.PEDI === "biological" /* BIOLOGICAL */ || motherType?.toLowerCase() === filter.PEDI.toLowerCase()) {
        const motherId = family.get("WIFE")?.toValue();
        const mother = motherId && this._gedcom?.indi(motherId);
        if (mother) {
          parents.append(mother);
        }
      }
    });
    return parents;
  }
  getSpousesFilteredByPartner(filter) {
    const spouses = new Individuals();
    const familiesOfSpouses = this.get("FAMS")?.toValueList();
    if (!familiesOfSpouses || !this.id) {
      return spouses;
    }
    const families = this.toFamilies(familiesOfSpouses);
    families.forEach((family) => {
      const spouseType = family.get("_SREL")?.toValue();
      if (!spouseType && filter.PART === "spouse" /* SPOUSE */ || spouseType?.toLowerCase() === filter.PART.toLowerCase()) {
        const spouse = family.getParents().copy().except(this).index(0);
        if (spouse) {
          spouses.append(spouse);
        }
      }
    });
    return spouses;
  }
  getBiologicalParents() {
    implemented();
    return this.getParentsFilteredByPedigree(BIOLOGICAL);
  }
  getAdoptedParents() {
    implemented();
    return this.getParentsFilteredByPedigree(ADOPTED);
  }
  getBirthParents() {
    implemented();
    return this.getParentsFilteredByPedigree(BIRTH);
  }
  getFosterParents() {
    implemented();
    return this.getParentsFilteredByPedigree(FOSTER);
  }
  getSealingParents() {
    implemented();
    return this.getParentsFilteredByPedigree(SEALING);
  }
  getStepParents() {
    implemented();
    return this.getParentsFilteredByPedigree(STEP);
  }
  getFathers() {
    implemented();
    return this.getParents().filter(MALE);
  }
  getBiologicalFathers() {
    implemented();
    return this.getParentsFilteredByPedigree(BIOLOGICAL).filter(
      MALE
    );
  }
  getAdoptedFathers() {
    implemented();
    return this.getParentsFilteredByPedigree(ADOPTED).filter(
      MALE
    );
  }
  getBirthFathers() {
    implemented();
    return this.getParentsFilteredByPedigree(BIRTH).filter(
      MALE
    );
  }
  getFosterFathers() {
    implemented();
    return this.getParentsFilteredByPedigree(FOSTER).filter(
      MALE
    );
  }
  getSealingFathers() {
    implemented();
    return this.getParentsFilteredByPedigree(SEALING).filter(
      MALE
    );
  }
  getStepFathers() {
    implemented();
    return this.getParentsFilteredByPedigree(STEP).filter(
      MALE
    );
  }
  getMothers() {
    implemented();
    return this.getParents().filter(FEMALE);
  }
  getBiologicalMothers() {
    implemented();
    return this.getParentsFilteredByPedigree(BIOLOGICAL).filter(
      FEMALE
    );
  }
  getAdoptedMothers() {
    implemented();
    return this.getParentsFilteredByPedigree(ADOPTED).filter(
      FEMALE
    );
  }
  getBirthMothers() {
    implemented();
    return this.getParentsFilteredByPedigree(BIRTH).filter(
      FEMALE
    );
  }
  getFosterMothers() {
    implemented();
    return this.getParentsFilteredByPedigree(FOSTER).filter(
      FEMALE
    );
  }
  getSealingMothers() {
    implemented();
    return this.getParentsFilteredByPedigree(SEALING).filter(
      FEMALE
    );
  }
  getStepMothers() {
    implemented();
    return this.getParentsFilteredByPedigree(STEP).filter(
      FEMALE
    );
  }
  getCoParents() {
    implemented();
    return this.getRelativesOnLevel(0);
  }
  getSpouses() {
    implemented();
    return this.getSpousesFilteredByPartner(SPOUSE);
  }
  getPartners() {
    implemented();
    return this.getSpousesFilteredByPartner(PARTNER);
  }
  getFriends() {
    implemented();
    return this.getSpousesFilteredByPartner(FRIEND);
  }
  getWives() {
    implemented();
    return this.getCoParents().filter(FEMALE);
  }
  getHusbands() {
    implemented();
    return this.getCoParents().filter(MALE);
  }
  getCousins() {
    implemented();
    return this.getRelativesOnDegree(1);
  }
  getGrandParents() {
    implemented();
    return this.getAscendants(2);
  }
  getGrandFathers() {
    implemented();
    return this.getGrandParents().filter(MALE);
  }
  getGrandMothers() {
    implemented();
    return this.getGrandParents().filter(FEMALE);
  }
  getGrandChildren() {
    implemented();
    return this.getDescendants(2);
  }
  getGrandSons() {
    implemented();
    return this.getGrandChildren().filter(MALE);
  }
  getGrandDaughters() {
    implemented();
    return this.getGrandChildren().filter(FEMALE);
  }
  getGreatGrandParents() {
    implemented();
    return this.getAscendants(3);
  }
  getGreatGrandFathers() {
    implemented();
    return this.getGreatGrandParents().filter(MALE);
  }
  getGreatGrandMothers() {
    implemented();
    return this.getGreatGrandParents().filter(FEMALE);
  }
  getGreatGrandChildren() {
    implemented();
    return this.getDescendants(3);
  }
  getGreatGrandSons() {
    implemented();
    return this.getGreatGrandChildren().filter(MALE);
  }
  getGreatGrandDaughters() {
    implemented();
    return this.getGreatGrandChildren().filter(FEMALE);
  }
  getAuncles() {
    implemented();
    return this.getParents().getSiblings();
  }
  getAunts() {
    implemented();
    return this.getAuncles().filter(FEMALE);
  }
  getUncles() {
    implemented();
    return this.getAuncles().filter(MALE);
  }
  getNiblings() {
    implemented();
    return this.getSiblings().getChildren();
  }
  getNieces() {
    implemented();
    return this.getNiblings().filter(FEMALE);
  }
  getNephews() {
    implemented();
    return this.getNiblings().filter(MALE);
  }
  getParentsInLaw() {
    implemented();
    return this.getCoParents().getParents();
  }
  getFathersInLaw() {
    implemented();
    return this.getParentsInLaw().filter(MALE);
  }
  getMothersInLaw() {
    implemented();
    return this.getParentsInLaw().filter(FEMALE);
  }
  getSiblingsInLaw() {
    implemented();
    return this.getCoParents().copy().getSiblings().copy().merge(this.getSiblings().getCoParents());
  }
  getBrothersInLaw() {
    implemented();
    return this.getSiblingsInLaw().filter(MALE);
  }
  getSistersInLaw() {
    implemented();
    return this.getSiblingsInLaw().filter(FEMALE);
  }
  getChildrenInLaw() {
    implemented();
    return this.getChildren().getCoParents();
  }
  getSonsInLaw() {
    implemented();
    return this.getChildrenInLaw().filter(MALE);
  }
  getDaughtersInLaw() {
    implemented();
    return this.getChildrenInLaw().filter(FEMALE);
  }
  // These are generated automatically
  get2ndCousins() {
    return new Individuals();
  }
  get2ndGreatGrandParents() {
    return new Individuals();
  }
  get2ndGreatGrandChildren() {
    return new Individuals();
  }
  get3rdCousins() {
    return new Individuals();
  }
  get3rdGreatGrandParents() {
    return new Individuals();
  }
  get3rdGreatGrandChildren() {
    return new Individuals();
  }
  get4thCousins() {
    return new Individuals();
  }
  get4thGreatGrandParents() {
    return new Individuals();
  }
  get4thGreatGrandChildren() {
    return new Individuals();
  }
  get5thCousins() {
    return new Individuals();
  }
  get5thGreatGrandParents() {
    return new Individuals();
  }
  get5thGreatGrandChildren() {
    return new Individuals();
  }
  get6thCousins() {
    return new Individuals();
  }
  get6thGreatGrandParents() {
    return new Individuals();
  }
  get6thGreatGrandChildren() {
    return new Individuals();
  }
  get7thCousins() {
    return new Individuals();
  }
  get7thGreatGrandParents() {
    return new Individuals();
  }
  get7thGreatGrandChildren() {
    return new Individuals();
  }
  get8thCousins() {
    return new Individuals();
  }
  get8thGreatGrandParents() {
    return new Individuals();
  }
  get8thGreatGrandChildren() {
    return new Individuals();
  }
  get9thCousins() {
    return new Individuals();
  }
  get9thGreatGrandParents() {
    return new Individuals();
  }
  get9thGreatGrandChildren() {
    return new Individuals();
  }
};
var generateFunctions = () => {
  const levels = [2, 3, 4, 5, 6, 7, 8, 9];
  const types = [
    ["Cousins", 0, "degree"],
    ["GreatGrandParents", 2, "level"],
    ["GreatGrandChildren", -2, "level"]
  ];
  levels.forEach((level) => {
    types.forEach(([type, starting, direction]) => {
      let validLevel;
      if (level === 2) {
        validLevel = "2nd";
      } else if (level === 3) {
        validLevel = "3rd";
      } else {
        validLevel = `${level}th`;
      }
      Indi.prototype[`get${validLevel}${type}`] = function(filter) {
        if (direction === "level" && starting < 0) {
          return this.getRelativesOnLevel(-level + starting, filter);
        } else if (direction === "level" && starting > 0) {
          return this.getRelativesOnLevel(level + starting, filter);
        }
        return this.getRelativesOnDegree(level + starting);
      };
    });
  });
};
generateFunctions();
var createIndi = (gedcom, id, main, parent) => {
  return createProxy(
    new Indi(gedcom, id, main, parent)
  );
};

// src/utils/name-formatter.ts
var nameFormatter = (indiName, settings, letterOnAll = true, indi, debug) => {
  const {
    nameOrder = "first-last",
    maxGivennames = 0,
    maxSurnames = 0
  } = settings ?? {};
  const rawName = indiName instanceof Indi ? indiName?.NAME : createCommonName(void 0, void 0, indi).name(indiName);
  const name = rawName?.index(0);
  let givenname = name?.GIVN?.toValueList().first()?.value || "";
  let surname = name?.SURN?.toValueList().first()?.value || "";
  let suffix = name?.NSFX?.toValueList().first()?.value || "";
  const displayedName = name?.DISPLAY?.toValueList().first()?.value || "";
  if (!givenname && !surname) {
    givenname = displayedName;
  }
  if (maxGivennames > 0) {
    givenname = givenname.split(" ").slice(0, maxGivennames).join(" ");
  }
  if (maxSurnames > 0) {
    surname = surname.split(" ").slice(0, maxSurnames).join(" ");
  }
  const inOrder = [
    suffix,
    ...nameOrder === "last-first" ? [surname, givenname] : [givenname, surname]
  ];
  const lName = (letterOnAll ? [inOrder[1], inOrder[2]].filter(Boolean).join(" ") : inOrder[1]).toLowerCase();
  const firstLetter = lName.match(/^(dzs|cs|dz|gy|ly|ny|sz|ty|zs|\w)/i)?.[0] ?? lName.substring(0, 1) ?? "";
  const validFirstLetter = `${firstLetter.substring(0, 1).toUpperCase()}${firstLetter.substring(1)}`;
  return { suffix, givenname, surname, inOrder, letter: validFirstLetter };
};

// src/utils/ordinalize.ts
var ordinalize = (n) => {
  const num = Math.round(n);
  const numString = num.toString();
  if (Math.floor(num / 10) % 10 === 1) {
    return numString + "th";
  }
  switch (num % 10) {
    case 1:
      return numString + "st";
    case 2:
      return numString + "nd";
    case 3:
      return numString + "rd";
    default:
      return numString + "th";
  }
};
var hungarianOrdinalize = (n) => {
  const numberPrefix = [
    "",
    "",
    "k\xE9t",
    "h\xE1rom",
    "n\xE9gy",
    "\xF6t",
    "hat",
    "h\xE9t",
    "nyolc",
    "kilenc"
  ];
  const ordinals = [
    [
      "",
      ["els\u0151", "egyed"],
      ["m\xE1sod", "ketted"],
      "harmad",
      "negyed",
      "\xF6t\xF6d",
      "hatod",
      "heted",
      "nyolcad",
      "kilenced"
    ],
    [
      "",
      ["tized", "tizen"],
      ["huszad", "huszon"],
      "harmincad",
      "negyvened",
      "\xF6tvened",
      "hatvanad",
      "hetvened",
      "nyolcvanad",
      "kilencvened"
    ],
    "sz\xE1zad",
    ["ezred", "ezer"]
  ];
  const parts = [];
  let fractionIndex = 0;
  let fractions = 1;
  let nr = Math.abs(n);
  let part;
  do {
    const nextFractions = fractions * 10;
    const index = fractions > 1 ? nr % nextFractions / fractions : nr % nextFractions;
    const ordinalList = ordinals[fractionIndex++];
    if (typeof ordinalList === "string") {
      part = `${numberPrefix[index]}${ordinalList}`;
    } else if (ordinalList.length === 2) {
      part = ordinalList;
    } else {
      part = ordinalList?.[index];
    }
    if (Array.isArray(part)) {
      if (fractions === 1) {
        part = part[nr < 10 ? 0 : 1];
      } else {
        part = part[n % 10 === 0 ? 0 : 1];
      }
    } else if (fractions > 1 && n % 10 !== 0) {
      part = part?.substring(0, part.length - 2);
    }
    if (part !== void 0) {
      parts.unshift(part);
    }
    fractions = nextFractions;
    nr = Math.floor(nr / nextFractions) * nextFractions;
    if (nr < nextFractions) {
      break;
    }
  } while (part !== void 0);
  return parts.join("");
};

// src/kinship-translator/kinship-translator.basic.ts
var KinshipTranslatorBasic = class {
  constructor(path, displayName = "givenname") {
    this.isOfSpouse = false;
    this.isSpouseOf = false;
    this._path = path;
    this.displayName = displayName;
  }
  get path() {
    return this._path;
  }
  get path0() {
    return this.path?.[0];
  }
  get path1() {
    return this.path?.[1];
  }
  get pathM() {
    return this.path?.[this.path.length - 2];
  }
  get pathN() {
    return this.path?.[this.path.length - 1];
  }
  get person1() {
    if (this.isOfSpouse) {
      return this.path1?.indi;
    }
    return this.path0?.indi;
  }
  get person2() {
    return this.path1?.indi;
  }
  get personM() {
    return this.pathM?.indi;
  }
  get personN() {
    if (this.isSpouseOf) {
      return this.pathM?.indi;
    }
    return this.pathN?.indi;
  }
  indirect() {
    return "";
  }
  removal() {
    return "";
  }
  parent() {
    return "";
  }
  child() {
    return "";
  }
  sibling() {
    return "";
  }
  spouse() {
    return "";
  }
  ofSpouse(relation) {
    return relation ?? "";
  }
  spouseOf(relation) {
    return relation ?? "";
  }
  relationType(relation) {
    return relation ?? "";
  }
  of(relation) {
    return relation ?? "";
  }
  inLaw(relation) {
    return relation ?? "";
  }
};

// src/kinship-translator/patterns.hu.ts
var parentRelationsHu = {
  step: "mostoha",
  foster: "nevelt",
  birth: "v\xE9rszerinti",
  biological: "v\xE9rszerinti",
  adopted: "\xF6r\xF6kbefogadott"
};
var nominativus = {
  apa: { nominativus: "apa", dativus: "apj\xE1nak", possessivus: "apja" },
  anya: { nominativus: "anya", dativus: "anyj\xE1nak", possessivus: "anyja" },
  ap\u00F3s: { nominativus: "ap\xF3s", dativus: "ap\xF3s\xE1nak", possessivus: "ap\xF3sa" },
  any\u00F3s: {
    nominativus: "any\xF3s",
    dativus: "any\xF3s\xE1nak",
    possessivus: "any\xF3sa"
  },
  v\u0151: { nominativus: "v\u0151", dativus: "vej\xE9nek", possessivus: "veje" },
  meny: { nominativus: "meny", dativus: "meny\xE9nek", possessivus: "menye" },
  s\u00F3gor: {
    nominativus: "s\xF3gor",
    dativus: "s\xF3gor\xE1nak",
    possessivus: "s\xF3gora"
  },
  s\u00F3gorn\u0151: {
    nominativus: "s\xF3gorn\u0151",
    dativus: "s\xF3gorn\u0151j\xE9nek",
    possessivus: "s\xF3gorn\u0151je"
  },
  b\u00E1csi: {
    nominativus: "b\xE1csi",
    dativus: "b\xE1tyj\xE1nak",
    possessivus: "b\xE1tyja"
  },
  n\u00E9ni: { nominativus: "n\xE9ni", dativus: "n\xE9nj\xE9nek", possessivus: "n\xE9nje" },
  testv\u00E9r: {
    nominativus: "testv\xE9r",
    dativus: "testv\xE9r\xE9nek",
    possessivus: "testv\xE9re"
  },
  \u00F6cs: { nominativus: "\xF6cs", dativus: "\xF6ccs\xE9nek", possessivus: "\xF6ccse" },
  h\u00FAg: { nominativus: "h\xFAg", dativus: "h\xFAg\xE1nak", possessivus: "h\xFAga" },
  f\u00E9rj: { nominativus: "f\xE9rj", dativus: "f\xE9rj\xE9nek", possessivus: "f\xE9rje" },
  feles\u00E9g: {
    nominativus: "feles\xE9g",
    dativus: "feles\xE9g\xE9nek",
    possessivus: "feles\xE9ge"
  },
  h\u00E1zast\u00E1rs: {
    nominativus: "h\xE1zast\xE1rs",
    dativus: "h\xE1zast\xE1rs\xE1nak",
    possessivus: "h\xE1zast\xE1rsa"
  },
  unoka: {
    nominativus: "unoka",
    dativus: "unok\xE1j\xE1nak",
    possessivus: "unok\xE1ja"
  },
  gyermek: {
    nominativus: "gyermek",
    dativus: "gyermek\xE9nek",
    possessivus: "gyermeke"
  },
  sz\u00FCl\u0151: {
    nominativus: "sz\xFCl\u0151",
    dativus: "sz\xFCl\u0151j\xE9nek",
    possessivus: "sz\xFCl\u0151je"
  }
};
var dativus = Object.values(nominativus).reduce((acc, val) => {
  acc[val.dativus] = val;
  return acc;
}, {});
var possessivus = Object.values(nominativus).reduce((acc, val) => {
  acc[val.possessivus] = val;
  return acc;
}, {});
var casesHu = {
  nominativus,
  dativus,
  possessivus
};
var InLawsHu = {
  "gyermek feles\xE9ge": "meny",
  "gyermek f\xE9rje": "v\u0151",
  [`(f\xE9rj|feles\xE9g) (?<mod1>(${Object.values(parentRelationsHu).join(
    "|"
  )}) )?anyja`]: "any\xF3s",
  [`(f\xE9rj|feles\xE9g) (?<mod1>(${Object.values(parentRelationsHu).join(
    "|"
  )}) )?apja`]: "ap\xF3s",
  [`(f\xE9rj|feles\xE9g) (?<mod1>(${Object.values(parentRelationsHu).join(
    "|"
  )}) )?(f\xE9l)?testv\xE9re`]: (indi) => {
    if (indi?.isMale()) {
      return "s\xF3gor";
    }
    if (indi?.isFemale()) {
      return "s\xF3gorn\u0151";
    }
  },
  [`(f\xE9l)?testv\xE9r f\xE9rje`]: "s\xF3gor",
  [`(f\xE9l)?testv\xE9r feles\xE9ge`]: "s\xF3gorn\u0151"
};

// src/kinship-translator/kinship-translator.hu.ts
var KinshipTranslatorHU = class extends KinshipTranslatorBasic {
  directPrefix(l) {
    const level = l === void 0 ? Math.abs(this.pathN?.level ?? 0) : l;
    if (level <= 1) {
      return "";
    }
    if (level === 2) {
      return "nagy";
    }
    if (level === 3) {
      return "d\xE9d";
    }
    if (level === 4) {
      return "\xFCk";
    }
    if (level === 5) {
      return "sz\xE9p";
    }
    if (level === 6) {
      return "\u0151s";
    }
    return `${level - 5}. \u0151s`;
  }
  caseReplacer(relation, to, from = "nominativus", mod = "$") {
    let newRelation = relation;
    Object.keys(casesHu[from]).forEach((c) => {
      newRelation = newRelation.replace(
        new RegExp(`(?<word1>${c}(?<mod>${mod}))`),
        (m, ...rest) => {
          const groups = rest[rest.length - 1];
          const word = groups?.word1 || groups?.word2;
          if (!word) {
            return m;
          }
          const toReplace = casesHu[from][word.trim()][to];
          return toReplace ? `${toReplace}${groups.mod || ""}` : m;
        }
      );
    });
    return newRelation;
  }
  indirect() {
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (degree === 1) {
      return `unokatestv\xE9r`;
    }
    return `${hungarianOrdinalize(degree)}unokatestv\xE9r`;
  }
  removal() {
    const level = this.pathN?.level ?? 0;
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (!degree || Math.abs(level) >= 1 && Math.abs(level) <= 6) {
      return level < 0 ? this.nibling() : this.auncle();
    }
    return `${hungarianOrdinalize(
      degree
    )}unokatestv\xE9r ${level}x elt\xE1vol\xEDtva`;
  }
  auncle() {
    const level = this.pathN?.level ?? 0;
    const degree = Math.abs(this.pathN?.degree ?? 0);
    let degreeString = degree > 1 ? hungarianOrdinalize(degree) : "";
    if (degreeString) {
      degreeString = `${degreeString}-`;
    }
    let prefix = this.directPrefix();
    const origPrefix = prefix;
    if (level === 2) {
      prefix = "nagy-";
    }
    if (this.personN?.isMale()) {
      return `${degreeString}${prefix}nagyb\xE1csi`;
    }
    if (this.personN?.isFemale()) {
      return `${degreeString}${prefix}nagyn\xE9ni`;
    }
    return `${origPrefix}sz\xFCl\u0151 ${degreeString ? `${degreeString}unoka` : ""}testv\xE9re`;
  }
  nibling() {
    const level = Math.abs(this.pathN?.level ?? 0);
    const degree = Math.abs(this.pathN?.degree ?? 0);
    const ordinalized = degree > 1 ? hungarianOrdinalize(degree) : "";
    let degreeString = ordinalized;
    if (degreeString) {
      degreeString = `${degreeString}-`;
    }
    const prefix = this.directPrefix(level + 1);
    const origPrefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${degreeString}${level === 1 ? origPrefix : prefix}unoka\xF6cs`;
    }
    if (this.personN?.isFemale()) {
      return `${degreeString}${level === 1 ? origPrefix : prefix}unokah\xFAg`;
    }
    if (level <= 1) {
      return `${ordinalized ? `${ordinalized}unoka` : ""}testv\xE9r gyermeke`;
    }
    return `${ordinalized ? `${ordinalized}unoka` : ""}testv\xE9r ${origPrefix.replace(/nagy$/, "")}unok\xE1ja`;
  }
  parent() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}apa`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}anya`;
    }
    return `${prefix}sz\xFCl\u0151`;
  }
  child() {
    const level = Math.abs(this.pathN?.level ?? 0);
    if (level <= 1) {
      return "gyermek";
    }
    const prefix = this.directPrefix();
    return `${prefix.replace(/nagy$/, "")}unoka`;
  }
  sibling() {
    const parents1 = this.person1?.getBiologicalParents();
    const parentsN = this.personN?.getBiologicalParents();
    const inter = parents1?.intersection(parentsN);
    if (!inter || inter.length < 2) {
      return "f\xE9ltestv\xE9r";
    }
    return "testv\xE9r";
  }
  spouse() {
    if (this.personN?.isMale()) {
      return "f\xE9rj";
    }
    if (this.personN?.isFemale()) {
      return "feles\xE9g";
    }
    return "h\xE1zast\xE1rs";
  }
  ofSpouse(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[1].indi;
    let spouseType = "h\xE1zast\xE1rs";
    if (spouse?.isMale()) {
      spouseType = "f\xE9rj";
    }
    if (spouse?.isFemale()) {
      spouseType = "feles\xE9g";
    }
    return `${spouseType} ${this.caseReplacer(relation, "possessivus")}`;
  }
  spouseOf(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[this.path.length - 1].indi;
    let spouseType = "h\xE1zast\xE1rsa";
    if (spouse?.isMale()) {
      spouseType = "f\xE9rje";
    }
    if (spouse?.isFemale()) {
      spouseType = "feles\xE9ge";
    }
    return `${this.caseReplacer(
      relation,
      "nominativus",
      "possessivus"
    )} ${spouseType}`;
  }
  of(relation) {
    const formattedName = nameFormatter(this.path?.[0]?.indi, {
      nameOrder: "last-first"
    });
    let name = formattedName.inOrder.filter(Boolean).join(" ");
    if (this.displayName === "none") {
      name = "";
    } else if (this.displayName === "givenname") {
      name = formattedName.givenname;
    } else if (this.displayName === "surname") {
      name = `${formattedName.surname}`;
    }
    const isSpouse = this.path?.[1].kinship === "spouse" || this.path?.[this.path.length - 1].kinship === "spouse";
    if (!name || !relation) {
      return relation ?? "";
    }
    if (isSpouse) {
      return `${name} ${this.caseReplacer(
        this.caseReplacer(relation, "dativus", "nominativus", " "),
        "possessivus"
      )}`;
    }
    return `${name} ${this.caseReplacer(
      this.caseReplacer(relation, "possessivus", "nominativus", " "),
      "possessivus"
    )}`;
  }
  inLaw(relation) {
    let inLawRelation = relation;
    Object.keys(InLawsHu).find((pattern) => {
      const regex = new RegExp(
        `(?<space> )(?<word1>${pattern})$|^(?<word2>${pattern.replace(
          "?<mod1>",
          "?<mod2>"
        )})$`
      );
      if (relation?.match(regex)) {
        inLawRelation = relation?.replace(regex, (m, ...rest) => {
          const groups = rest[rest.length - 1];
          const word = groups?.word1 || groups?.word2;
          const mod = groups?.mod1 || groups?.mod2;
          if (!word) {
            return m;
          }
          let toReplace = InLawsHu[word.trim()] || InLawsHu[pattern];
          if (typeof toReplace === "function") {
            toReplace = toReplace(this.personN) ?? "";
          }
          return toReplace ? `${groups.space || ""}${mod || ""}${toReplace}` : m;
        });
        return true;
      }
      return false;
    });
    return inLawRelation ?? "";
  }
  relationType(relation) {
    if (!relation || !this.pathN?.relation) {
      return relation ?? "";
    }
    return `${parentRelationsHu[this.pathN.relation]} ${relation}`;
  }
};

// src/kinship-translator/patterns.en.ts
var InLawsEn = {
  "spouse of (child|daughter|son)": "child-in-law",
  "wife of (child|daughter|son)": "daughter-in-law",
  "husband of (child|daughter|son)": "son-in-law",
  "mother of (spouse|wife|husband)": "mother-in-law",
  "father of (spouse|wife|husband)": "father-in-law",
  "parent of (spouse|wife|husband)": "parent-in-law",
  "wife of (half-)?(sibling|sister|brother)": "sister-in-law",
  "husband of (half-)?(sibling|sister|brother)": "brother-in-law",
  "spouse of (half-)?(sibling|sister|brother)": "sibling-in-law",
  "(half-)?sister of (spouse|wife|husband)": "sister-in-law",
  "(half-)?brother of (spouse|wife|husband)": "brother-in-law",
  "(half-)?sibling of (spouse|wife|husband)": "sibling-in-law"
};

// src/kinship-translator/kinship-translator.en.ts
var KinshipTranslatorEn = class extends KinshipTranslatorBasic {
  directPrefix() {
    const level = Math.abs(this.pathN?.level ?? 0);
    if (level <= 1) {
      return "";
    }
    if (level === 2) {
      return "grand";
    }
    if (level === 3) {
      return "great-grand";
    }
    return `${ordinalize(level - 2)} great-grand`;
  }
  indirect() {
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (degree === 1) {
      return `cousin`;
    }
    return `${ordinalize(degree)} cousin`;
  }
  removal() {
    const level = this.pathN?.level ?? 0;
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (!degree) {
      return level < 0 ? this.nibling() : this.auncle();
    }
    return `${ordinalize(degree)} cousin ${level}x removed`;
  }
  auncle() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}uncle`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}aunt`;
    }
    return `${prefix}auncle`;
  }
  nibling() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}nephew`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}niece`;
    }
    return `${prefix}nibling`;
  }
  parent() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}father`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}mother`;
    }
    return `${prefix}parent`;
  }
  child() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}son`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}daughter`;
    }
    return `${prefix}child`;
  }
  sibling() {
    const parents1 = this.person1?.getBiologicalParents();
    const parentsN = this.personN?.getBiologicalParents();
    const inter = parents1?.intersection(parentsN);
    let prefix = "";
    if (!inter || inter.length < 2) {
      prefix = "half-";
    }
    if (this.personN?.isMale()) {
      return `${prefix}brother`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}sister`;
    }
    return `${prefix}sibling`;
  }
  spouse() {
    if (this.personN?.isMale()) {
      return "husband";
    }
    if (this.personN?.isFemale()) {
      return "wife";
    }
    return "spouse";
  }
  ofSpouse(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[1].indi;
    let spouseType = "spouse";
    if (spouse?.isMale()) {
      spouseType = "husband";
    }
    if (spouse?.isFemale()) {
      spouseType = "wife";
    }
    return `${relation} of ${spouseType}`;
  }
  spouseOf(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[this.path.length - 1].indi;
    let spouseType = "spouse";
    if (spouse?.isMale()) {
      spouseType = "husband";
    }
    if (spouse?.isFemale()) {
      spouseType = "wife";
    }
    return `${spouseType} of ${relation}`;
  }
  of(relation) {
    const formattedName = nameFormatter(this.path?.[0]?.indi, {
      nameOrder: "first-last"
    });
    let name = formattedName.inOrder.filter(Boolean).join(" ");
    if (this.displayName === "none") {
      name = "";
    } else if (this.displayName === "givenname") {
      name = formattedName.givenname;
    } else if (this.displayName === "surname") {
      name = `${formattedName.surname}`;
    }
    if (!name || !relation) {
      return relation ?? "";
    }
    if (relation.includes(" of ")) {
      return relation.replace(" of ", ` of ${name}'s `);
    }
    return `${name}'s ${relation}`;
  }
  inLaw(relation) {
    let inLawRelation = relation;
    Object.keys(InLawsEn).find((pattern) => {
      const regex = new RegExp(
        `(?<space> )(?<word1>${pattern})$|^(?<word2>${pattern})$`
      );
      if (relation?.match(regex)) {
        inLawRelation = relation?.replace(regex, (m, ...rest) => {
          const groups = rest[rest.length - 1];
          const word = groups?.word1 || groups?.word2;
          if (!word) {
            return m;
          }
          const toReplace = InLawsEn[word.trim()] || InLawsEn[pattern];
          return toReplace ? `${groups.space || ""}${toReplace}` : m;
        });
        return true;
      }
      return false;
    });
    return inLawRelation ?? "";
  }
  relationType(relation) {
    if (!relation || !this.pathN?.relation) {
      return relation ?? "";
    }
    return `${this.pathN.relation} ${relation}`;
  }
};

// src/kinship-translator/patterns.de.ts
var InLawsDe = {
  "Ehepartner von (kind|tochter|sohn)": "Schwiegerkind",
  "Ehefrau von (kind|tochter|sohn)": "Schwiegertochter",
  "Ehemann von (kind|tochter|sohn)": "Schwiegersohn",
  "mutter des (Ehepartner|Ehefrau|Ehemann)s": "Schwiegermutter",
  "vater des (Ehepartner|Ehefrau|Ehemann)s": "Schwiegervater",
  "elternteil des (Ehepartner|Ehefrau|Ehemann)s": "Schwiegerelternteil",
  "Ehefrau von (Halb)?(geschwister|schwester|bruder)": "Schw\xE4gerin",
  "Ehemann von (Halb)?(geschwister|schwester|bruder)": "Schwager",
  "Ehepartner von (Halb)?(geschwister|schwester|bruder)": "Schwager/Schw\xE4gerin",
  "(Halb)?schwester des (Ehepartner|Ehefrau|Ehemann)s": "Schw\xE4gerin",
  "(Halb)?bruder des (Ehepartner|Ehefrau|Ehemann)s": "Schwager",
  "(Halb)?geschwister des (Ehepartner|Ehefrau|Ehemann)s": "Schwager/Schw\xE4gerin"
};

// src/kinship-translator/kinship-translator.de.ts
var KinshipTranslatorDe = class extends KinshipTranslatorBasic {
  directPrefix() {
    const level = Math.abs(this.pathN?.level ?? 0);
    if (level <= 1) {
      return "";
    }
    if (level === 2) {
      return "Gro\xDF";
    }
    if (level === 3) {
      return "Urgro\xDF";
    }
    return `Ur${level - 2}-gro\xDF`;
  }
  indirect() {
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (degree === 1) {
      return `Cousin`;
    }
    return `Cousin ${degree}. Grades`;
  }
  removal() {
    const level = this.pathN?.level ?? 0;
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (!degree) {
      return level < 0 ? this.nibling() : this.auncle();
    }
    return `Cousin ${degree}. Grades ${Math.abs(level)}\xD7 entfernt`;
  }
  auncle() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}onkel`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}tante`;
    }
    return `${prefix}onkel/tante`;
  }
  nibling() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}neffe`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}nichte`;
    }
    return `${prefix}neffe/nichte`;
  }
  parent() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}vater`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}mutter`;
    }
    return `${prefix}elternteil`;
  }
  child() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}sohn`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}tochter`;
    }
    return `${prefix}kind`;
  }
  sibling() {
    const parents1 = this.person1?.getBiologicalParents();
    const parentsN = this.personN?.getBiologicalParents();
    const inter = parents1?.intersection(parentsN);
    let prefix = "";
    if (!inter || inter.length < 2) {
      prefix = "Halb";
    }
    if (this.personN?.isMale()) {
      return `${prefix}bruder`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}schwester`;
    }
    return `${prefix}geschwister`;
  }
  spouse() {
    if (this.personN?.isMale()) {
      return "Ehemann";
    }
    if (this.personN?.isFemale()) {
      return "Ehefrau";
    }
    return "Ehepartner";
  }
  ofSpouse(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[1].indi;
    let spouseType = "Ehepartner";
    if (spouse?.isMale()) {
      spouseType = "Ehemann";
    }
    if (spouse?.isFemale()) {
      spouseType = "Ehefrau";
    }
    return `${relation} des ${spouseType}s`;
  }
  spouseOf(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[this.path.length - 1].indi;
    let spouseType = "Ehepartner";
    if (spouse?.isMale()) {
      spouseType = "Ehemann";
    }
    if (spouse?.isFemale()) {
      spouseType = "Ehefrau";
    }
    return `${spouseType} von ${relation}`;
  }
  of(relation) {
    const formattedName = nameFormatter(this.path?.[0]?.indi, {
      nameOrder: "first-last"
    });
    let name = formattedName.inOrder.filter(Boolean).join(" ");
    if (this.displayName === "none") {
      name = "";
    } else if (this.displayName === "givenname") {
      name = formattedName.givenname;
    } else if (this.displayName === "surname") {
      name = `${formattedName.surname}`;
    }
    if (!name || !relation) {
      return relation ?? "";
    }
    if (relation.includes(" des ") || relation.includes(" von ")) {
      return relation.replace(/ (des|von) /, ` $1 ${name}s `);
    }
    return `${name}s ${relation}`;
  }
  inLaw(relation) {
    let inLawRelation = relation;
    Object.keys(InLawsDe).find((pattern) => {
      const regex = new RegExp(
        `(?<space> )(?<word1>${pattern})$|^(?<word2>${pattern})$`
      );
      if (relation?.match(regex)) {
        inLawRelation = relation?.replace(regex, (m, ...rest) => {
          const groups = rest[rest.length - 1];
          const word = groups?.word1 || groups?.word2;
          if (!word) {
            return m;
          }
          const toReplace = InLawsDe[word.trim()] || InLawsDe[pattern];
          return toReplace ? `${groups.space || ""}${toReplace}` : m;
        });
        return true;
      }
      return false;
    });
    return inLawRelation ?? "";
  }
  relationType(relation) {
    if (!relation || !this.pathN?.relation) {
      return relation ?? "";
    }
    return `${this.pathN.relation} ${relation}`;
  }
};

// src/kinship-translator/patterns.fr.ts
var InLawsFr = {
  "conjoint de (enfant|fille|fils)": "beau-fils/belle-fille",
  "\xE9pouse de (enfant|fille|fils)": "belle-fille",
  "mari de (enfant|fille|fils)": "gendre",
  "m\xE8re du (conjoint|\xE9pouse|mari)": "belle-m\xE8re",
  "p\xE8re du (conjoint|\xE9pouse|mari)": "beau-p\xE8re",
  "parent du (conjoint|\xE9pouse|mari)": "beau-parent",
  "\xE9pouse de (demi-)?(fr\xE8re/s\u0153ur|s\u0153ur|fr\xE8re)": "belle-s\u0153ur",
  "mari de (demi-)?(fr\xE8re/s\u0153ur|s\u0153ur|fr\xE8re)": "beau-fr\xE8re",
  "conjoint de (demi-)?(fr\xE8re/s\u0153ur|s\u0153ur|fr\xE8re)": "beau-fr\xE8re/belle-s\u0153ur",
  "(demi-)?s\u0153ur du (conjoint|\xE9pouse|mari)": "belle-s\u0153ur",
  "(demi-)?fr\xE8re du (conjoint|\xE9pouse|mari)": "beau-fr\xE8re",
  "(demi-)?fr\xE8re/s\u0153ur du (conjoint|\xE9pouse|mari)": "beau-fr\xE8re/belle-s\u0153ur"
};

// src/kinship-translator/kinship-translator.fr.ts
var KinshipTranslatorFr = class extends KinshipTranslatorBasic {
  directPrefix() {
    const level = Math.abs(this.pathN?.level ?? 0);
    if (level <= 1) {
      return "";
    }
    if (level === 2) {
      return "grand-";
    }
    if (level === 3) {
      return "arri\xE8re-grand-";
    }
    return `arri\xE8re${level - 2}-grand-`;
  }
  indirect() {
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (degree === 1) {
      return `cousin`;
    }
    return `cousin au ${ordinalize(degree)} degr\xE9`;
  }
  removal() {
    const level = this.pathN?.level ?? 0;
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (!degree) {
      return level < 0 ? this.nibling() : this.auncle();
    }
    return `cousin au ${ordinalize(degree)} degr\xE9 ${Math.abs(level)}\xD7 \xE9loign\xE9`;
  }
  auncle() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}oncle`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}tante`;
    }
    return `${prefix}oncle/tante`;
  }
  nibling() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}neveu`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}ni\xE8ce`;
    }
    return `${prefix}neveu/ni\xE8ce`;
  }
  parent() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}p\xE8re`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}m\xE8re`;
    }
    return `${prefix}parent`;
  }
  child() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}fils`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}fille`;
    }
    return `${prefix}enfant`;
  }
  sibling() {
    const parents1 = this.person1?.getBiologicalParents();
    const parentsN = this.personN?.getBiologicalParents();
    const inter = parents1?.intersection(parentsN);
    let prefix = "";
    if (!inter || inter.length < 2) {
      prefix = "demi-";
    }
    if (this.personN?.isMale()) {
      return `${prefix}fr\xE8re`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}s\u0153ur`;
    }
    return `${prefix}fr\xE8re/s\u0153ur`;
  }
  spouse() {
    if (this.personN?.isMale()) {
      return "mari";
    }
    if (this.personN?.isFemale()) {
      return "\xE9pouse";
    }
    return "conjoint";
  }
  ofSpouse(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[1].indi;
    let spouseType = "conjoint";
    if (spouse?.isMale()) {
      spouseType = "mari";
    }
    if (spouse?.isFemale()) {
      spouseType = "\xE9pouse";
    }
    return `${relation} du ${spouseType}`;
  }
  spouseOf(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[this.path.length - 1].indi;
    let spouseType = "conjoint";
    if (spouse?.isMale()) {
      spouseType = "mari";
    }
    if (spouse?.isFemale()) {
      spouseType = "\xE9pouse";
    }
    return `${spouseType} de ${relation}`;
  }
  of(relation) {
    const formattedName = nameFormatter(this.path?.[0]?.indi, {
      nameOrder: "first-last"
    });
    let name = formattedName.inOrder.filter(Boolean).join(" ");
    if (this.displayName === "none") {
      name = "";
    } else if (this.displayName === "givenname") {
      name = formattedName.givenname;
    } else if (this.displayName === "surname") {
      name = `${formattedName.surname}`;
    }
    if (!name || !relation) {
      return relation ?? "";
    }
    if (relation.includes(" du ") || relation.includes(" de ")) {
      return relation.replace(/ (du|de) /, ` $1 ${name} `);
    }
    return `${relation} de ${name}`;
  }
  inLaw(relation) {
    let inLawRelation = relation;
    Object.keys(InLawsFr).find((pattern) => {
      const regex = new RegExp(
        `(?<space> )(?<word1>${pattern})$|^(?<word2>${pattern})$`
      );
      if (relation?.match(regex)) {
        inLawRelation = relation?.replace(regex, (m, ...rest) => {
          const groups = rest[rest.length - 1];
          const word = groups?.word1 || groups?.word2;
          if (!word) {
            return m;
          }
          const toReplace = InLawsFr[word.trim()] || InLawsFr[pattern];
          return toReplace ? `${groups.space || ""}${toReplace}` : m;
        });
        return true;
      }
      return false;
    });
    return inLawRelation ?? "";
  }
  relationType(relation) {
    if (!relation || !this.pathN?.relation) {
      return relation ?? "";
    }
    return `${this.pathN.relation} ${relation}`;
  }
};

// src/kinship-translator/patterns.es.ts
var InLawsEs = {
  "c\xF3nyuge de (hijo/hija|hija|hijo)": "yerno/nuera",
  "esposa de (hijo/hija|hija|hijo)": "nuera",
  "esposo de (hijo/hija|hija|hijo)": "yerno",
  "madre del (c\xF3nyuge|esposa|esposo)": "suegra",
  "padre del (c\xF3nyuge|esposa|esposo)": "suegro",
  "padre/madre del (c\xF3nyuge|esposa|esposo)": "suegro/suegra",
  "esposa de (medio )?(hermano/hermana|hermana|hermano)": "cu\xF1ada",
  "esposo de (medio )?(hermano/hermana|hermana|hermano)": "cu\xF1ado",
  "c\xF3nyuge de (medio )?(hermano/hermana|hermana|hermano)": "cu\xF1ado/cu\xF1ada",
  "(medio )?hermana del (c\xF3nyuge|esposa|esposo)": "cu\xF1ada",
  "(medio )?hermano del (c\xF3nyuge|esposa|esposo)": "cu\xF1ado",
  "(medio )?hermano/hermana del (c\xF3nyuge|esposa|esposo)": "cu\xF1ado/cu\xF1ada"
};

// src/kinship-translator/kinship-translator.es.ts
var KinshipTranslatorEs = class extends KinshipTranslatorBasic {
  directPrefix() {
    const level = Math.abs(this.pathN?.level ?? 0);
    if (level <= 1) {
      return "";
    }
    if (level === 2) {
      return "bisabuel";
    }
    if (level === 3) {
      return "tatarabuel";
    }
    return `t\xE1tara${level - 2}-abuel`;
  }
  indirect() {
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (degree === 1) {
      return `primo`;
    }
    return `primo ${ordinalize(degree)}`;
  }
  removal() {
    const level = this.pathN?.level ?? 0;
    let degree = Math.abs(this.pathN?.degree ?? 0);
    if (degree > 0) {
      degree = degree - 1;
    }
    if (!degree) {
      return level < 0 ? this.nibling() : this.auncle();
    }
    return `primo ${ordinalize(degree)} ${Math.abs(level)}\xD7 separado`;
  }
  auncle() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}t\xEDo`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}t\xEDa`;
    }
    return `${prefix}t\xEDo/t\xEDa`;
  }
  nibling() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}sobrino`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}sobrina`;
    }
    return `${prefix}sobrino/sobrina`;
  }
  parent() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}padre`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}madre`;
    }
    return `${prefix}padre/madre`;
  }
  child() {
    const prefix = this.directPrefix();
    if (this.personN?.isMale()) {
      return `${prefix}hijo`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}hija`;
    }
    return `${prefix}hijo/hija`;
  }
  sibling() {
    const parents1 = this.person1?.getBiologicalParents();
    const parentsN = this.personN?.getBiologicalParents();
    const inter = parents1?.intersection(parentsN);
    let prefix = "";
    if (!inter || inter.length < 2) {
      prefix = "medio ";
    }
    if (this.personN?.isMale()) {
      return `${prefix}hermano`;
    }
    if (this.personN?.isFemale()) {
      return `${prefix}hermana`;
    }
    return `${prefix}hermano/hermana`;
  }
  spouse() {
    if (this.personN?.isMale()) {
      return "esposo";
    }
    if (this.personN?.isFemale()) {
      return "esposa";
    }
    return "c\xF3nyuge";
  }
  ofSpouse(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[1].indi;
    let spouseType = "c\xF3nyuge";
    if (spouse?.isMale()) {
      spouseType = "esposo";
    }
    if (spouse?.isFemale()) {
      spouseType = "esposa";
    }
    return `${relation} del ${spouseType}`;
  }
  spouseOf(relation) {
    if (!relation) {
      return "";
    }
    const spouse = this.path?.[this.path.length - 1].indi;
    let spouseType = "c\xF3nyuge";
    if (spouse?.isMale()) {
      spouseType = "esposo";
    }
    if (spouse?.isFemale()) {
      spouseType = "esposa";
    }
    return `${spouseType} de ${relation}`;
  }
  of(relation) {
    const formattedName = nameFormatter(this.path?.[0]?.indi, {
      nameOrder: "first-last"
    });
    let name = formattedName.inOrder.filter(Boolean).join(" ");
    if (this.displayName === "none") {
      name = "";
    } else if (this.displayName === "givenname") {
      name = formattedName.givenname;
    } else if (this.displayName === "surname") {
      name = `${formattedName.surname}`;
    }
    if (!name || !relation) {
      return relation ?? "";
    }
    if (relation.includes(" del ") || relation.includes(" de ")) {
      return relation.replace(/ (del|de) /, ` $1 ${name} `);
    }
    return `${relation} de ${name}`;
  }
  inLaw(relation) {
    let inLawRelation = relation;
    Object.keys(InLawsEs).find((pattern) => {
      const regex = new RegExp(
        `(?<space> )(?<word1>${pattern})$|^(?<word2>${pattern})$`
      );
      if (relation?.match(regex)) {
        inLawRelation = relation?.replace(regex, (m, ...rest) => {
          const groups = rest[rest.length - 1];
          const word = groups?.word1 || groups?.word2;
          if (!word) {
            return m;
          }
          const toReplace = InLawsEs[word.trim()] || InLawsEs[pattern];
          return toReplace ? `${groups.space || ""}${toReplace}` : m;
        });
        return true;
      }
      return false;
    });
    return inLawRelation ?? "";
  }
  relationType(relation) {
    if (!relation || !this.pathN?.relation) {
      return relation ?? "";
    }
    return `${this.pathN.relation} ${relation}`;
  }
};

// src/kinship-translator/kinship-translator.ts
var KinshipTranslator = class {
  constructor(person1, person2, lang = "en", entirePath, displayName = "givenname") {
    this.displayName = displayName;
    this.lang = lang;
    this.entirePath = entirePath;
    this.person1 = person1;
    this.person2 = typeof person2 === "string" ? this.person1.getGedcom()?.indi(person2) : person2;
    this.path = this.person1.path(this.person2);
    if (translators_exports[lang]) {
      this.translator = new translators_exports[lang](
        this.path ?? [],
        displayName
      );
    }
    if (!this.translator) {
      this.translator = new KinshipTranslatorEn(this.path ?? [], displayName);
    }
  }
  get path0() {
    return this.path?.[0];
  }
  get path1() {
    return this.path?.[1];
  }
  get pathM() {
    return this.path?.[this.path.length - 2];
  }
  get pathN() {
    return this.path?.[this.path.length - 1];
  }
  direct() {
    if (this.pathN?.level === 0 && this.pathN.degree === 0) {
      return this.translator?.spouse();
    }
    if (this.pathN?.level === 0 && this.pathN.degree === 1) {
      return this.translator?.sibling();
    }
    if ((this.pathN?.level ?? 0) > 0 && this.pathN?.degree === 0) {
      return this.translator?.parent();
    }
    if ((this.pathN?.level ?? 0) < 0 && this.pathN?.degree === 0) {
      return this.translator?.child();
    }
  }
  indirect() {
    return this.translator?.indirect();
  }
  removal() {
    return this.translator?.removal();
  }
  translate(showMainPerson) {
    if (this.entirePath) {
      const path = this.path?.map((path2, index) => ({
        id: path2.indi.id,
        gen: path2.level,
        absolute: this.person1.kinship(
          path2.indi.id,
          showMainPerson,
          this.lang,
          false,
          this.displayName
        ),
        relative: this.path?.[index - 1]?.indi.kinship(
          path2.indi.id,
          showMainPerson,
          this.lang,
          false,
          this.displayName
        )
      }));
      return path;
    }
    const person1 = this.path0;
    const person2 = this.path1;
    const personM = this.pathM;
    const personN = this.pathN;
    if (!person1 || !person2 || !personM || !personN) {
      return "";
    }
    if (person2.kinship === "spouse" && person2.indi.id !== personN.indi.id) {
      this.translator && (this.translator.isOfSpouse = true);
    } else if (personN.kinship === "spouse" && person2.indi.id !== personN.indi.id) {
      this.translator && (this.translator.isSpouseOf = true);
    }
    let relation;
    if (personN.degree === 0 || personN.degree === 1 && personN.level === 0) {
      relation = this.direct();
    } else if (personN.level === 0) {
      relation = this.indirect();
    } else if (personN.level !== 0) {
      relation = this.removal();
    }
    relation = this.translator?.relationType(relation);
    if (this.translator?.isOfSpouse) {
      relation = this.translator?.ofSpouse(relation);
    } else if (this.translator?.isSpouseOf) {
      relation = this.translator?.spouseOf(relation);
    }
    relation = this.translator?.inLaw(relation);
    if (showMainPerson && relation) {
      relation = this.translator?.of(relation);
    }
    return relation ?? "";
  }
};

// src/factories/kinship-factory.ts
var KinshipTranslatorClass = KinshipTranslator;
var setKinshipTranslatorClass = (TranslatorClass) => {
  KinshipTranslatorClass = TranslatorClass;
};
var getKinshipTranslatorClass = () => {
  return KinshipTranslatorClass;
};
var resetKinshipTranslatorClass = () => {
  KinshipTranslatorClass = KinshipTranslator;
};

export { getCacheManagerFactory, getDateLocale, getI18n, getKinshipTranslatorClass, getPlaceParserProvider, getPlaceTranslatorProvider, i18n, resetCacheManagerFactory, resetDateLocaleProvider, resetI18nProvider, resetKinshipTranslatorClass, resetPlaceParserProvider, resetPlaceTranslatorProvider, setCacheManagerFactory, setDateLocaleProvider, setI18nProvider, setKinshipTranslatorClass, setPlaceParserProvider, setPlaceTranslatorProvider };
